// @mysten/deepbook-v3@0.26.6/dist/cjs/contracts/deepbook/account.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var account_exports = {};
__export(account_exports, {
  Account: () => Account,
  activeStake: () => activeStake,
  createdProposal: () => createdProposal,
  inactiveStake: () => inactiveStake,
  makerVolume: () => makerVolume,
  openOrders: () => openOrders,
  settledBalances: () => settledBalances,
  takerVolume: () => takerVolume,
  totalVolume: () => totalVolume,
  votedProposal: () => votedProposal
});
module.exports = __toCommonJS(account_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var vec_set = __toESM(require("./deps/sui/vec_set.js"));
var balances = __toESM(require("./balances.js"));
const $moduleName = "@deepbook/core::account";
const Account = new import_utils.MoveStruct({
  name: `${$moduleName}::Account`,
  fields: {
    epoch: import_bcs.bcs.u64(),
    open_orders: vec_set.VecSet(import_bcs.bcs.u128()),
    taker_volume: import_bcs.bcs.u128(),
    maker_volume: import_bcs.bcs.u128(),
    active_stake: import_bcs.bcs.u64(),
    inactive_stake: import_bcs.bcs.u64(),
    created_proposal: import_bcs.bcs.bool(),
    voted_proposal: import_bcs.bcs.option(import_bcs.bcs.Address),
    unclaimed_rebates: balances.Balances,
    settled_balances: balances.Balances,
    owed_balances: balances.Balances
  }
});
function openOrders(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::account::Account`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "account",
    function: "open_orders",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function takerVolume(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::account::Account`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "account",
    function: "taker_volume",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function makerVolume(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::account::Account`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "account",
    function: "maker_volume",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function totalVolume(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::account::Account`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "account",
    function: "total_volume",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function activeStake(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::account::Account`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "account",
    function: "active_stake",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function inactiveStake(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::account::Account`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "account",
    function: "inactive_stake",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function createdProposal(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::account::Account`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "account",
    function: "created_proposal",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function votedProposal(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::account::Account`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "account",
    function: "voted_proposal",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function settledBalances(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::account::Account`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "account",
    function: "settled_balances",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
