// @mysten/deepbook-v3@0.26.6/dist/cjs/contracts/deepbook/balance_manager.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var balance_manager_exports = {};
__export(balance_manager_exports, {
  BalanceEvent: () => BalanceEvent,
  BalanceKey: () => BalanceKey,
  BalanceManager: () => BalanceManager,
  BalanceManagerEvent: () => BalanceManagerEvent,
  DepositCap: () => DepositCap,
  TradeCap: () => TradeCap,
  TradeProof: () => TradeProof,
  WithdrawCap: () => WithdrawCap,
  _new: () => _new,
  balance: () => balance,
  deposit: () => deposit,
  depositWithCap: () => depositWithCap,
  generateProofAsOwner: () => generateProofAsOwner,
  generateProofAsTrader: () => generateProofAsTrader,
  id: () => id,
  mintDepositCap: () => mintDepositCap,
  mintTradeCap: () => mintTradeCap,
  mintWithdrawCap: () => mintWithdrawCap,
  newWithCustomOwner: () => newWithCustomOwner,
  newWithCustomOwnerAndCaps: () => newWithCustomOwnerAndCaps,
  newWithOwner: () => newWithOwner,
  owner: () => owner,
  revokeTradeCap: () => revokeTradeCap,
  validateProof: () => validateProof,
  withdraw: () => withdraw,
  withdrawAll: () => withdrawAll,
  withdrawWithCap: () => withdrawWithCap
});
module.exports = __toCommonJS(balance_manager_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var object = __toESM(require("./deps/sui/object.js"));
var bag = __toESM(require("./deps/sui/bag.js"));
var vec_set = __toESM(require("./deps/sui/vec_set.js"));
var type_name = __toESM(require("./deps/std/type_name.js"));
const $moduleName = "@deepbook/core::balance_manager";
const BalanceManager = new import_utils.MoveStruct({
  name: `${$moduleName}::BalanceManager`,
  fields: {
    id: object.UID,
    owner: import_bcs.bcs.Address,
    balances: bag.Bag,
    allow_listed: vec_set.VecSet(import_bcs.bcs.Address)
  }
});
const BalanceManagerEvent = new import_utils.MoveStruct({
  name: `${$moduleName}::BalanceManagerEvent`,
  fields: {
    balance_manager_id: import_bcs.bcs.Address,
    owner: import_bcs.bcs.Address
  }
});
const BalanceEvent = new import_utils.MoveStruct({
  name: `${$moduleName}::BalanceEvent`,
  fields: {
    balance_manager_id: import_bcs.bcs.Address,
    asset: type_name.TypeName,
    amount: import_bcs.bcs.u64(),
    deposit: import_bcs.bcs.bool()
  }
});
const BalanceKey = new import_utils.MoveStruct({
  name: `${$moduleName}::BalanceKey`,
  fields: {
    dummy_field: import_bcs.bcs.bool()
  }
});
const TradeCap = new import_utils.MoveStruct({
  name: `${$moduleName}::TradeCap`,
  fields: {
    id: object.UID,
    balance_manager_id: import_bcs.bcs.Address
  }
});
const DepositCap = new import_utils.MoveStruct({
  name: `${$moduleName}::DepositCap`,
  fields: {
    id: object.UID,
    balance_manager_id: import_bcs.bcs.Address
  }
});
const WithdrawCap = new import_utils.MoveStruct({
  name: `${$moduleName}::WithdrawCap`,
  fields: {
    id: object.UID,
    balance_manager_id: import_bcs.bcs.Address
  }
});
const TradeProof = new import_utils.MoveStruct({
  name: `${$moduleName}::TradeProof`,
  fields: {
    balance_manager_id: import_bcs.bcs.Address,
    trader: import_bcs.bcs.Address
  }
});
function _new(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "new"
  });
}
function newWithOwner(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["address"];
  const parameterNames = ["Owner"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "new_with_owner",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function newWithCustomOwner(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["address"];
  const parameterNames = ["owner"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "new_with_custom_owner",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function newWithCustomOwnerAndCaps(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["address"];
  const parameterNames = ["owner"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "new_with_custom_owner_and_caps",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function balance(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::balance_manager::BalanceManager`];
  const parameterNames = ["balanceManager"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "balance",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function mintTradeCap(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::balance_manager::BalanceManager`];
  const parameterNames = ["balanceManager"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "mint_trade_cap",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function mintDepositCap(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::balance_manager::BalanceManager`];
  const parameterNames = ["balanceManager"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "mint_deposit_cap",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function mintWithdrawCap(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::balance_manager::BalanceManager`];
  const parameterNames = ["balanceManager"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "mint_withdraw_cap",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function revokeTradeCap(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::balance_manager::BalanceManager`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID"
  ];
  const parameterNames = ["balanceManager", "tradeCapId"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "revoke_trade_cap",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function generateProofAsOwner(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::balance_manager::BalanceManager`];
  const parameterNames = ["balanceManager"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "generate_proof_as_owner",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function generateProofAsTrader(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::balance_manager::BalanceManager`,
    `${packageAddress}::balance_manager::TradeCap`
  ];
  const parameterNames = ["balanceManager", "tradeCap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "generate_proof_as_trader",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function deposit(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::balance_manager::BalanceManager`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${options.typeArguments[0]}>`
  ];
  const parameterNames = ["balanceManager", "coin"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "deposit",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function depositWithCap(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::balance_manager::BalanceManager`,
    `${packageAddress}::balance_manager::DepositCap`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${options.typeArguments[0]}>`
  ];
  const parameterNames = ["balanceManager", "depositCap", "coin"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "deposit_with_cap",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function withdrawWithCap(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::balance_manager::BalanceManager`,
    `${packageAddress}::balance_manager::WithdrawCap`,
    "u64"
  ];
  const parameterNames = ["balanceManager", "withdrawCap", "withdrawAmount"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "withdraw_with_cap",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function withdraw(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::balance_manager::BalanceManager`,
    "u64"
  ];
  const parameterNames = ["balanceManager", "withdrawAmount"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "withdraw",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function withdrawAll(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::balance_manager::BalanceManager`];
  const parameterNames = ["balanceManager"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "withdraw_all",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function validateProof(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::balance_manager::BalanceManager`,
    `${packageAddress}::balance_manager::TradeProof`
  ];
  const parameterNames = ["balanceManager", "proof"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "validate_proof",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function owner(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::balance_manager::BalanceManager`];
  const parameterNames = ["balanceManager"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "owner",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function id(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::balance_manager::BalanceManager`];
  const parameterNames = ["balanceManager"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "id",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
