// @mysten/deepbook-v3@0.26.6/dist/cjs/contracts/deepbook/big_vector.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var big_vector_exports = {};
__export(big_vector_exports, {
  BigVector: () => BigVector,
  Slice: () => Slice,
  SliceRef: () => SliceRef
});
module.exports = __toCommonJS(big_vector_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var object = __toESM(require("./deps/sui/object.js"));
const $moduleName = "@deepbook/core::big_vector";
const BigVector = new import_utils.MoveStruct({
  name: `${$moduleName}::BigVector`,
  fields: {
    id: object.UID,
    /** How deep the tree structure is. */
    depth: import_bcs.bcs.u8(),
    /**
     * Total number of elements that this vector contains, not including gaps in the
     * vector.
     */
    length: import_bcs.bcs.u64(),
    /** Max size of leaf nodes (counted in number of elements, `E`). */
    max_slice_size: import_bcs.bcs.u64(),
    /** Max size of interior nodes (counted in number of children). */
    max_fan_out: import_bcs.bcs.u64(),
    /** ID of the tree's root structure. Value of `NO_SLICE` means there's no root. */
    root_id: import_bcs.bcs.u64(),
    /** The last node ID that was allocated. */
    last_id: import_bcs.bcs.u64()
  }
});
function Slice(...typeParameters) {
  return new import_utils.MoveStruct({
    name: `${$moduleName}::Slice<${typeParameters[0].name}>`,
    fields: {
      /** Previous node in the intrusive doubly-linked list data structure. */
      prev: import_bcs.bcs.u64(),
      /** Next node in the intrusive doubly-linked list data structure. */
      next: import_bcs.bcs.u64(),
      keys: import_bcs.bcs.vector(import_bcs.bcs.u128()),
      vals: import_bcs.bcs.vector(typeParameters[0])
    }
  });
}
const SliceRef = new import_utils.MoveStruct({
  name: `${$moduleName}::SliceRef`,
  fields: {
    ix: import_bcs.bcs.u64()
  }
});
