// @mysten/deepbook-v3@0.26.6/dist/cjs/contracts/deepbook/book.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var book_exports = {};
__export(book_exports, {
  Book: () => Book
});
module.exports = __toCommonJS(book_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var big_vector = __toESM(require("./big_vector.js"));
const $moduleName = "@deepbook/core::book";
const Book = new import_utils.MoveStruct({
  name: `${$moduleName}::Book`,
  fields: {
    tick_size: import_bcs.bcs.u64(),
    lot_size: import_bcs.bcs.u64(),
    min_size: import_bcs.bcs.u64(),
    bids: big_vector.BigVector,
    asks: big_vector.BigVector,
    next_bid_order_id: import_bcs.bcs.u64(),
    next_ask_order_id: import_bcs.bcs.u64()
  }
});
