// @mysten/deepbook-v3@0.26.6/dist/cjs/contracts/deepbook/constants.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  cancelMaker: () => cancelMaker,
  cancelTaker: () => cancelTaker,
  canceled: () => canceled,
  currentVersion: () => currentVersion,
  deepUnit: () => deepUnit,
  defaultAdditionalTakerFee: () => defaultAdditionalTakerFee,
  defaultEwmaAlpha: () => defaultEwmaAlpha,
  defaultStakeRequired: () => defaultStakeRequired,
  defaultZScoreThreshold: () => defaultZScoreThreshold,
  ewmaDfKey: () => ewmaDfKey,
  expired: () => expired,
  feeIsDeep: () => feeIsDeep,
  feePenaltyMultiplier: () => feePenaltyMultiplier,
  fillOrKill: () => fillOrKill,
  filled: () => filled,
  floatScaling: () => floatScaling,
  floatScalingU128: () => floatScalingU128,
  half: () => half,
  immediateOrCancel: () => immediateOrCancel,
  live: () => live,
  maxFanOut: () => maxFanOut,
  maxFills: () => maxFills,
  maxOpenOrders: () => maxOpenOrders,
  maxPrice: () => maxPrice,
  maxRestriction: () => maxRestriction,
  maxSliceSize: () => maxSliceSize,
  maxU128: () => maxU128,
  maxU64: () => maxU64,
  minPrice: () => minPrice,
  noRestriction: () => noRestriction,
  partiallyFilled: () => partiallyFilled,
  phaseOutEpochs: () => phaseOutEpochs,
  poolCreationFee: () => poolCreationFee,
  postOnly: () => postOnly,
  selfMatchingAllowed: () => selfMatchingAllowed
});
module.exports = __toCommonJS(constants_exports);
function currentVersion(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "current_version"
  });
}
function poolCreationFee(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "pool_creation_fee"
  });
}
function floatScaling(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "float_scaling"
  });
}
function floatScalingU128(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "float_scaling_u128"
  });
}
function maxU64(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "max_u64"
  });
}
function maxU128(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "max_u128"
  });
}
function noRestriction(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "no_restriction"
  });
}
function immediateOrCancel(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "immediate_or_cancel"
  });
}
function fillOrKill(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "fill_or_kill"
  });
}
function postOnly(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "post_only"
  });
}
function maxRestriction(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "max_restriction"
  });
}
function live(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "live"
  });
}
function partiallyFilled(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "partially_filled"
  });
}
function filled(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "filled"
  });
}
function canceled(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "canceled"
  });
}
function expired(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "expired"
  });
}
function selfMatchingAllowed(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "self_matching_allowed"
  });
}
function cancelTaker(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "cancel_taker"
  });
}
function cancelMaker(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "cancel_maker"
  });
}
function minPrice(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "min_price"
  });
}
function maxPrice(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "max_price"
  });
}
function phaseOutEpochs(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "phase_out_epochs"
  });
}
function defaultStakeRequired(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "default_stake_required"
  });
}
function half(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "half"
  });
}
function feeIsDeep(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "fee_is_deep"
  });
}
function deepUnit(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "deep_unit"
  });
}
function maxFills(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "max_fills"
  });
}
function maxOpenOrders(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "max_open_orders"
  });
}
function maxSliceSize(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "max_slice_size"
  });
}
function maxFanOut(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "max_fan_out"
  });
}
function feePenaltyMultiplier(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "fee_penalty_multiplier"
  });
}
function defaultEwmaAlpha(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "default_ewma_alpha"
  });
}
function defaultZScoreThreshold(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "default_z_score_threshold"
  });
}
function defaultAdditionalTakerFee(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "default_additional_taker_fee"
  });
}
function ewmaDfKey(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "ewma_df_key"
  });
}
