// @mysten/deepbook-v3@0.26.6/dist/cjs/contracts/deepbook/deep_price.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var deep_price_exports = {};
__export(deep_price_exports, {
  DeepPrice: () => DeepPrice,
  OrderDeepPrice: () => OrderDeepPrice,
  Price: () => Price,
  PriceAdded: () => PriceAdded,
  assetIsBase: () => assetIsBase,
  deepPerAsset: () => deepPerAsset
});
module.exports = __toCommonJS(deep_price_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
const $moduleName = "@deepbook/core::deep_price";
const Price = new import_utils.MoveStruct({
  name: `${$moduleName}::Price`,
  fields: {
    conversion_rate: import_bcs.bcs.u64(),
    timestamp: import_bcs.bcs.u64()
  }
});
const PriceAdded = new import_utils.MoveStruct({
  name: `${$moduleName}::PriceAdded`,
  fields: {
    conversion_rate: import_bcs.bcs.u64(),
    timestamp: import_bcs.bcs.u64(),
    is_base_conversion: import_bcs.bcs.bool(),
    reference_pool: import_bcs.bcs.Address,
    target_pool: import_bcs.bcs.Address
  }
});
const DeepPrice = new import_utils.MoveStruct({
  name: `${$moduleName}::DeepPrice`,
  fields: {
    base_prices: import_bcs.bcs.vector(Price),
    cumulative_base: import_bcs.bcs.u64(),
    quote_prices: import_bcs.bcs.vector(Price),
    cumulative_quote: import_bcs.bcs.u64()
  }
});
const OrderDeepPrice = new import_utils.MoveStruct({
  name: `${$moduleName}::OrderDeepPrice`,
  fields: {
    asset_is_base: import_bcs.bcs.bool(),
    deep_per_asset: import_bcs.bcs.u64()
  }
});
function assetIsBase(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::deep_price::OrderDeepPrice`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "deep_price",
    function: "asset_is_base",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function deepPerAsset(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::deep_price::OrderDeepPrice`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "deep_price",
    function: "deep_per_asset",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
