// @mysten/deepbook-v3@0.26.6/dist/cjs/contracts/deepbook/governance.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var governance_exports = {};
__export(governance_exports, {
  Governance: () => Governance,
  Proposal: () => Proposal,
  TradeParamsUpdateEvent: () => TradeParamsUpdateEvent
});
module.exports = __toCommonJS(governance_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var vec_map = __toESM(require("./deps/sui/vec_map.js"));
var trade_params = __toESM(require("./trade_params.js"));
const $moduleName = "@deepbook/core::governance";
const Proposal = new import_utils.MoveStruct({
  name: `${$moduleName}::Proposal`,
  fields: {
    taker_fee: import_bcs.bcs.u64(),
    maker_fee: import_bcs.bcs.u64(),
    stake_required: import_bcs.bcs.u64(),
    votes: import_bcs.bcs.u64()
  }
});
const Governance = new import_utils.MoveStruct({
  name: `${$moduleName}::Governance`,
  fields: {
    /** Tracks refreshes. */
    epoch: import_bcs.bcs.u64(),
    /** If Pool is whitelisted. */
    whitelisted: import_bcs.bcs.bool(),
    /** If Pool is stable or volatile. */
    stable: import_bcs.bcs.bool(),
    /** List of proposals for the current epoch. */
    proposals: vec_map.VecMap(import_bcs.bcs.Address, Proposal),
    /** Trade parameters for the current epoch. */
    trade_params: trade_params.TradeParams,
    /** Trade parameters for the next epoch. */
    next_trade_params: trade_params.TradeParams,
    /** All voting power from the current stakes. */
    voting_power: import_bcs.bcs.u64(),
    /** Quorum for the current epoch. */
    quorum: import_bcs.bcs.u64()
  }
});
const TradeParamsUpdateEvent = new import_utils.MoveStruct({
  name: `${$moduleName}::TradeParamsUpdateEvent`,
  fields: {
    taker_fee: import_bcs.bcs.u64(),
    maker_fee: import_bcs.bcs.u64(),
    stake_required: import_bcs.bcs.u64()
  }
});
