// @mysten/deepbook-v3@0.26.6/dist/cjs/contracts/deepbook/history.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var history_exports = {};
__export(history_exports, {
  EpochData: () => EpochData,
  History: () => History,
  Volumes: () => Volumes
});
module.exports = __toCommonJS(history_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var balances = __toESM(require("./balances.js"));
var trade_params = __toESM(require("./trade_params.js"));
var table = __toESM(require("./deps/sui/table.js"));
const $moduleName = "@deepbook/core::history";
const Volumes = new import_utils.MoveStruct({
  name: `${$moduleName}::Volumes`,
  fields: {
    total_volume: import_bcs.bcs.u128(),
    total_staked_volume: import_bcs.bcs.u128(),
    total_fees_collected: balances.Balances,
    historic_median: import_bcs.bcs.u128(),
    trade_params: trade_params.TradeParams
  }
});
const History = new import_utils.MoveStruct({
  name: `${$moduleName}::History`,
  fields: {
    epoch: import_bcs.bcs.u64(),
    epoch_created: import_bcs.bcs.u64(),
    volumes: Volumes,
    historic_volumes: table.Table,
    balance_to_burn: import_bcs.bcs.u64()
  }
});
const EpochData = new import_utils.MoveStruct({
  name: `${$moduleName}::EpochData`,
  fields: {
    epoch: import_bcs.bcs.u64(),
    pool_id: import_bcs.bcs.Address,
    total_volume: import_bcs.bcs.u128(),
    total_staked_volume: import_bcs.bcs.u128(),
    base_fees_collected: import_bcs.bcs.u64(),
    quote_fees_collected: import_bcs.bcs.u64(),
    deep_fees_collected: import_bcs.bcs.u64(),
    historic_median: import_bcs.bcs.u128(),
    taker_fee: import_bcs.bcs.u64(),
    maker_fee: import_bcs.bcs.u64(),
    stake_required: import_bcs.bcs.u64()
  }
});
