// @mysten/deepbook-v3@0.26.6/dist/cjs/contracts/deepbook/math.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var math_exports = {};
__export(math_exports, {
  div: () => div,
  divRoundUp: () => divRoundUp,
  divU128: () => divU128,
  isPowerOfTen: () => isPowerOfTen,
  median: () => median,
  mul: () => mul,
  mulRoundUp: () => mulRoundUp,
  mulU128: () => mulU128,
  sqrt: () => sqrt
});
module.exports = __toCommonJS(math_exports);
var import_utils = require("../utils/index.js");
function mul(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["u64", "u64"];
  const parameterNames = ["x", "y"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "math",
    function: "mul",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function mulU128(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["u128", "u128"];
  const parameterNames = ["x", "y"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "math",
    function: "mul_u128",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function mulRoundUp(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["u64", "u64"];
  const parameterNames = ["x", "y"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "math",
    function: "mul_round_up",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function div(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["u64", "u64"];
  const parameterNames = ["x", "y"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "math",
    function: "div",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function divU128(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["u128", "u128"];
  const parameterNames = ["x", "y"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "math",
    function: "div_u128",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function divRoundUp(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["u64", "u64"];
  const parameterNames = ["x", "y"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "math",
    function: "div_round_up",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function median(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["vector<u128>"];
  const parameterNames = ["v"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "math",
    function: "median",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function sqrt(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["u64", "u64"];
  const parameterNames = ["x", "precision"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "math",
    function: "sqrt",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function isPowerOfTen(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["u64"];
  const parameterNames = ["n"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "math",
    function: "is_power_of_ten",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
