// @mysten/deepbook-v3@0.26.6/dist/cjs/contracts/deepbook/order_query.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var order_query_exports = {};
__export(order_query_exports, {
  OrderPage: () => OrderPage,
  hasNextPage: () => hasNextPage,
  iterOrders: () => iterOrders,
  orders: () => orders
});
module.exports = __toCommonJS(order_query_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var order = __toESM(require("./order.js"));
const $moduleName = "@deepbook/core::order_query";
const OrderPage = new import_utils.MoveStruct({
  name: `${$moduleName}::OrderPage`,
  fields: {
    orders: import_bcs.bcs.vector(order.Order),
    has_next_page: import_bcs.bcs.bool()
  }
});
function iterOrders(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::pool::Pool<${options.typeArguments[0]}, ${options.typeArguments[1]}>`,
    "0x0000000000000000000000000000000000000000000000000000000000000001::option::Option<u128>",
    "0x0000000000000000000000000000000000000000000000000000000000000001::option::Option<u128>",
    "0x0000000000000000000000000000000000000000000000000000000000000001::option::Option<u64>",
    "u64",
    "bool"
  ];
  const parameterNames = [
    "self",
    "startOrderId",
    "endOrderId",
    "minExpireTimestamp",
    "limit",
    "bids"
  ];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_query",
    function: "iter_orders",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function orders(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_query::OrderPage`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_query",
    function: "orders",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function hasNextPage(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_query::OrderPage`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_query",
    function: "has_next_page",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
