// @mysten/deepbook-v3@0.26.6/dist/cjs/contracts/deepbook/order_info.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var order_info_exports = {};
__export(order_info_exports, {
  OrderExpired: () => OrderExpired,
  OrderFilled: () => OrderFilled,
  OrderInfo: () => OrderInfo,
  OrderPlaced: () => OrderPlaced,
  balanceManagerId: () => balanceManagerId,
  clientOrderId: () => clientOrderId,
  cumulativeQuoteQuantity: () => cumulativeQuoteQuantity,
  epoch: () => epoch,
  executedQuantity: () => executedQuantity,
  expireTimestamp: () => expireTimestamp,
  feeIsDeep: () => feeIsDeep,
  fillLimitReached: () => fillLimitReached,
  fills: () => fills,
  isBid: () => isBid,
  makerFees: () => makerFees,
  orderDeepPrice: () => orderDeepPrice,
  orderId: () => orderId,
  orderInserted: () => orderInserted,
  orderType: () => orderType,
  originalQuantity: () => originalQuantity,
  paidFees: () => paidFees,
  poolId: () => poolId,
  price: () => price,
  selfMatchingOption: () => selfMatchingOption,
  status: () => status,
  trader: () => trader
});
module.exports = __toCommonJS(order_info_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var deep_price = __toESM(require("./deep_price.js"));
var fill = __toESM(require("./fill.js"));
const $moduleName = "@deepbook/core::order_info";
const OrderInfo = new import_utils.MoveStruct({
  name: `${$moduleName}::OrderInfo`,
  fields: {
    pool_id: import_bcs.bcs.Address,
    order_id: import_bcs.bcs.u128(),
    balance_manager_id: import_bcs.bcs.Address,
    client_order_id: import_bcs.bcs.u64(),
    trader: import_bcs.bcs.Address,
    order_type: import_bcs.bcs.u8(),
    self_matching_option: import_bcs.bcs.u8(),
    price: import_bcs.bcs.u64(),
    is_bid: import_bcs.bcs.bool(),
    original_quantity: import_bcs.bcs.u64(),
    order_deep_price: deep_price.OrderDeepPrice,
    expire_timestamp: import_bcs.bcs.u64(),
    executed_quantity: import_bcs.bcs.u64(),
    cumulative_quote_quantity: import_bcs.bcs.u64(),
    fills: import_bcs.bcs.vector(fill.Fill),
    fee_is_deep: import_bcs.bcs.bool(),
    paid_fees: import_bcs.bcs.u64(),
    maker_fees: import_bcs.bcs.u64(),
    epoch: import_bcs.bcs.u64(),
    status: import_bcs.bcs.u8(),
    market_order: import_bcs.bcs.bool(),
    fill_limit_reached: import_bcs.bcs.bool(),
    order_inserted: import_bcs.bcs.bool(),
    timestamp: import_bcs.bcs.u64()
  }
});
const OrderFilled = new import_utils.MoveStruct({
  name: `${$moduleName}::OrderFilled`,
  fields: {
    pool_id: import_bcs.bcs.Address,
    maker_order_id: import_bcs.bcs.u128(),
    taker_order_id: import_bcs.bcs.u128(),
    maker_client_order_id: import_bcs.bcs.u64(),
    taker_client_order_id: import_bcs.bcs.u64(),
    price: import_bcs.bcs.u64(),
    taker_is_bid: import_bcs.bcs.bool(),
    taker_fee: import_bcs.bcs.u64(),
    taker_fee_is_deep: import_bcs.bcs.bool(),
    maker_fee: import_bcs.bcs.u64(),
    maker_fee_is_deep: import_bcs.bcs.bool(),
    base_quantity: import_bcs.bcs.u64(),
    quote_quantity: import_bcs.bcs.u64(),
    maker_balance_manager_id: import_bcs.bcs.Address,
    taker_balance_manager_id: import_bcs.bcs.Address,
    timestamp: import_bcs.bcs.u64()
  }
});
const OrderPlaced = new import_utils.MoveStruct({
  name: `${$moduleName}::OrderPlaced`,
  fields: {
    balance_manager_id: import_bcs.bcs.Address,
    pool_id: import_bcs.bcs.Address,
    order_id: import_bcs.bcs.u128(),
    client_order_id: import_bcs.bcs.u64(),
    trader: import_bcs.bcs.Address,
    price: import_bcs.bcs.u64(),
    is_bid: import_bcs.bcs.bool(),
    placed_quantity: import_bcs.bcs.u64(),
    expire_timestamp: import_bcs.bcs.u64(),
    timestamp: import_bcs.bcs.u64()
  }
});
const OrderExpired = new import_utils.MoveStruct({
  name: `${$moduleName}::OrderExpired`,
  fields: {
    balance_manager_id: import_bcs.bcs.Address,
    pool_id: import_bcs.bcs.Address,
    order_id: import_bcs.bcs.u128(),
    client_order_id: import_bcs.bcs.u64(),
    trader: import_bcs.bcs.Address,
    price: import_bcs.bcs.u64(),
    is_bid: import_bcs.bcs.bool(),
    original_quantity: import_bcs.bcs.u64(),
    base_asset_quantity_canceled: import_bcs.bcs.u64(),
    timestamp: import_bcs.bcs.u64()
  }
});
function poolId(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "pool_id",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function orderId(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "order_id",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function balanceManagerId(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "balance_manager_id",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function clientOrderId(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "client_order_id",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function trader(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "trader",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function orderType(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "order_type",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function selfMatchingOption(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "self_matching_option",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function price(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "price",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function isBid(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "is_bid",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function originalQuantity(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "original_quantity",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function orderDeepPrice(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "order_deep_price",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function expireTimestamp(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "expire_timestamp",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function executedQuantity(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "executed_quantity",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function cumulativeQuoteQuantity(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "cumulative_quote_quantity",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function fills(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "fills",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function feeIsDeep(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "fee_is_deep",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function paidFees(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "paid_fees",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function makerFees(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "maker_fees",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function epoch(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "epoch",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function status(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "status",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function fillLimitReached(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "fill_limit_reached",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function orderInserted(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "order_inserted",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
