// @mysten/deepbook-v3@0.26.6/dist/cjs/contracts/deepbook/order.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var order_exports = {};
__export(order_exports, {
  Order: () => Order,
  OrderCanceled: () => OrderCanceled,
  OrderModified: () => OrderModified,
  balanceManagerId: () => balanceManagerId,
  clientOrderId: () => clientOrderId,
  epoch: () => epoch,
  expireTimestamp: () => expireTimestamp,
  feeIsDeep: () => feeIsDeep,
  filledQuantity: () => filledQuantity,
  orderDeepPrice: () => orderDeepPrice,
  orderId: () => orderId,
  price: () => price,
  quantity: () => quantity,
  status: () => status
});
module.exports = __toCommonJS(order_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var deep_price = __toESM(require("./deep_price.js"));
const $moduleName = "@deepbook/core::order";
const Order = new import_utils.MoveStruct({
  name: `${$moduleName}::Order`,
  fields: {
    balance_manager_id: import_bcs.bcs.Address,
    order_id: import_bcs.bcs.u128(),
    client_order_id: import_bcs.bcs.u64(),
    quantity: import_bcs.bcs.u64(),
    filled_quantity: import_bcs.bcs.u64(),
    fee_is_deep: import_bcs.bcs.bool(),
    order_deep_price: deep_price.OrderDeepPrice,
    epoch: import_bcs.bcs.u64(),
    status: import_bcs.bcs.u8(),
    expire_timestamp: import_bcs.bcs.u64()
  }
});
const OrderCanceled = new import_utils.MoveStruct({
  name: `${$moduleName}::OrderCanceled`,
  fields: {
    balance_manager_id: import_bcs.bcs.Address,
    pool_id: import_bcs.bcs.Address,
    order_id: import_bcs.bcs.u128(),
    client_order_id: import_bcs.bcs.u64(),
    trader: import_bcs.bcs.Address,
    price: import_bcs.bcs.u64(),
    is_bid: import_bcs.bcs.bool(),
    original_quantity: import_bcs.bcs.u64(),
    base_asset_quantity_canceled: import_bcs.bcs.u64(),
    timestamp: import_bcs.bcs.u64()
  }
});
const OrderModified = new import_utils.MoveStruct({
  name: `${$moduleName}::OrderModified`,
  fields: {
    balance_manager_id: import_bcs.bcs.Address,
    pool_id: import_bcs.bcs.Address,
    order_id: import_bcs.bcs.u128(),
    client_order_id: import_bcs.bcs.u64(),
    trader: import_bcs.bcs.Address,
    price: import_bcs.bcs.u64(),
    is_bid: import_bcs.bcs.bool(),
    previous_quantity: import_bcs.bcs.u64(),
    filled_quantity: import_bcs.bcs.u64(),
    new_quantity: import_bcs.bcs.u64(),
    timestamp: import_bcs.bcs.u64()
  }
});
function balanceManagerId(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order::Order`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order",
    function: "balance_manager_id",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function orderId(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order::Order`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order",
    function: "order_id",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function clientOrderId(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order::Order`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order",
    function: "client_order_id",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function quantity(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order::Order`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order",
    function: "quantity",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function filledQuantity(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order::Order`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order",
    function: "filled_quantity",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function feeIsDeep(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order::Order`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order",
    function: "fee_is_deep",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function orderDeepPrice(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order::Order`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order",
    function: "order_deep_price",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function epoch(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order::Order`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order",
    function: "epoch",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function status(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order::Order`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order",
    function: "status",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function expireTimestamp(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order::Order`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order",
    function: "expire_timestamp",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function price(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order::Order`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order",
    function: "price",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
