// @mysten/deepbook-v3@0.26.6/dist/cjs/contracts/deepbook/registry.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var registry_exports = {};
__export(registry_exports, {
  DeepbookAdminCap: () => DeepbookAdminCap,
  PoolKey: () => PoolKey,
  REGISTRY: () => REGISTRY,
  Registry: () => Registry,
  RegistryInner: () => RegistryInner,
  StableCoinKey: () => StableCoinKey,
  addStablecoin: () => addStablecoin,
  disableVersion: () => disableVersion,
  enableVersion: () => enableVersion,
  isStablecoin: () => isStablecoin,
  removeStablecoin: () => removeStablecoin,
  setTreasuryAddress: () => setTreasuryAddress
});
module.exports = __toCommonJS(registry_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var object = __toESM(require("./deps/sui/object.js"));
var versioned = __toESM(require("./deps/sui/versioned.js"));
var vec_set = __toESM(require("./deps/sui/vec_set.js"));
var bag = __toESM(require("./deps/sui/bag.js"));
var type_name = __toESM(require("./deps/std/type_name.js"));
const $moduleName = "@deepbook/core::registry";
const REGISTRY = new import_utils.MoveStruct({
  name: `${$moduleName}::REGISTRY`,
  fields: {
    dummy_field: import_bcs.bcs.bool()
  }
});
const DeepbookAdminCap = new import_utils.MoveStruct({
  name: `${$moduleName}::DeepbookAdminCap`,
  fields: {
    id: object.UID
  }
});
const Registry = new import_utils.MoveStruct({
  name: `${$moduleName}::Registry`,
  fields: {
    id: object.UID,
    inner: versioned.Versioned
  }
});
const RegistryInner = new import_utils.MoveStruct({
  name: `${$moduleName}::RegistryInner`,
  fields: {
    allowed_versions: vec_set.VecSet(import_bcs.bcs.u64()),
    pools: bag.Bag,
    treasury_address: import_bcs.bcs.Address
  }
});
const PoolKey = new import_utils.MoveStruct({
  name: `${$moduleName}::PoolKey`,
  fields: {
    base: type_name.TypeName,
    quote: type_name.TypeName
  }
});
const StableCoinKey = new import_utils.MoveStruct({
  name: `${$moduleName}::StableCoinKey`,
  fields: {
    dummy_field: import_bcs.bcs.bool()
  }
});
function setTreasuryAddress(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::registry::Registry`,
    "address",
    `${packageAddress}::registry::DeepbookAdminCap`
  ];
  const parameterNames = ["self", "treasuryAddress", "Cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "registry",
    function: "set_treasury_address",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function enableVersion(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::registry::Registry`,
    "u64",
    `${packageAddress}::registry::DeepbookAdminCap`
  ];
  const parameterNames = ["self", "version", "Cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "registry",
    function: "enable_version",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function disableVersion(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::registry::Registry`,
    "u64",
    `${packageAddress}::registry::DeepbookAdminCap`
  ];
  const parameterNames = ["self", "version", "Cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "registry",
    function: "disable_version",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function addStablecoin(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::registry::Registry`,
    `${packageAddress}::registry::DeepbookAdminCap`
  ];
  const parameterNames = ["self", "Cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "registry",
    function: "add_stablecoin",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function removeStablecoin(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::registry::Registry`,
    `${packageAddress}::registry::DeepbookAdminCap`
  ];
  const parameterNames = ["self", "Cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "registry",
    function: "remove_stablecoin",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function isStablecoin(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::registry::Registry`,
    "0x0000000000000000000000000000000000000000000000000000000000000001::type_name::TypeName"
  ];
  const parameterNames = ["self", "stableType"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "registry",
    function: "is_stablecoin",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
