// @mysten/deepbook-v3@0.26.6/dist/cjs/contracts/deepbook/state.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var state_exports = {};
__export(state_exports, {
  ProposalEvent: () => ProposalEvent,
  RebateEvent: () => RebateEvent,
  RebateEventV2: () => RebateEventV2,
  StakeEvent: () => StakeEvent,
  State: () => State,
  VoteEvent: () => VoteEvent
});
module.exports = __toCommonJS(state_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var table = __toESM(require("./deps/sui/table.js"));
var history = __toESM(require("./history.js"));
var governance = __toESM(require("./governance.js"));
var balances = __toESM(require("./balances.js"));
const $moduleName = "@deepbook/core::state";
const State = new import_utils.MoveStruct({
  name: `${$moduleName}::State`,
  fields: {
    accounts: table.Table,
    history: history.History,
    governance: governance.Governance
  }
});
const StakeEvent = new import_utils.MoveStruct({
  name: `${$moduleName}::StakeEvent`,
  fields: {
    pool_id: import_bcs.bcs.Address,
    balance_manager_id: import_bcs.bcs.Address,
    epoch: import_bcs.bcs.u64(),
    amount: import_bcs.bcs.u64(),
    stake: import_bcs.bcs.bool()
  }
});
const ProposalEvent = new import_utils.MoveStruct({
  name: `${$moduleName}::ProposalEvent`,
  fields: {
    pool_id: import_bcs.bcs.Address,
    balance_manager_id: import_bcs.bcs.Address,
    epoch: import_bcs.bcs.u64(),
    taker_fee: import_bcs.bcs.u64(),
    maker_fee: import_bcs.bcs.u64(),
    stake_required: import_bcs.bcs.u64()
  }
});
const VoteEvent = new import_utils.MoveStruct({
  name: `${$moduleName}::VoteEvent`,
  fields: {
    pool_id: import_bcs.bcs.Address,
    balance_manager_id: import_bcs.bcs.Address,
    epoch: import_bcs.bcs.u64(),
    from_proposal_id: import_bcs.bcs.option(import_bcs.bcs.Address),
    to_proposal_id: import_bcs.bcs.Address,
    stake: import_bcs.bcs.u64()
  }
});
const RebateEventV2 = new import_utils.MoveStruct({
  name: `${$moduleName}::RebateEventV2`,
  fields: {
    pool_id: import_bcs.bcs.Address,
    balance_manager_id: import_bcs.bcs.Address,
    epoch: import_bcs.bcs.u64(),
    claim_amount: balances.Balances
  }
});
const RebateEvent = new import_utils.MoveStruct({
  name: `${$moduleName}::RebateEvent`,
  fields: {
    pool_id: import_bcs.bcs.Address,
    balance_manager_id: import_bcs.bcs.Address,
    epoch: import_bcs.bcs.u64(),
    claim_amount: import_bcs.bcs.u64()
  }
});
