// @mysten/deepbook-v3@0.26.6/dist/cjs/contracts/deepbook/trade_params.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var trade_params_exports = {};
__export(trade_params_exports, {
  TradeParams: () => TradeParams
});
module.exports = __toCommonJS(trade_params_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
const $moduleName = "@deepbook/core::trade_params";
const TradeParams = new import_utils.MoveStruct({
  name: `${$moduleName}::TradeParams`,
  fields: {
    taker_fee: import_bcs.bcs.u64(),
    maker_fee: import_bcs.bcs.u64(),
    stake_required: import_bcs.bcs.u64()
  }
});
