// @mysten/deepbook-v3@0.26.6/dist/cjs/contracts/deepbook/vault.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var vault_exports = {};
__export(vault_exports, {
  FlashLoan: () => FlashLoan,
  FlashLoanBorrowed: () => FlashLoanBorrowed,
  Vault: () => Vault
});
module.exports = __toCommonJS(vault_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var balance = __toESM(require("./deps/sui/balance.js"));
var type_name = __toESM(require("./deps/std/type_name.js"));
const $moduleName = "@deepbook/core::vault";
const Vault = new import_utils.MoveStruct({
  name: `${$moduleName}::Vault`,
  fields: {
    base_balance: balance.Balance,
    quote_balance: balance.Balance,
    deep_balance: balance.Balance
  }
});
const FlashLoan = new import_utils.MoveStruct({
  name: `${$moduleName}::FlashLoan`,
  fields: {
    pool_id: import_bcs.bcs.Address,
    borrow_quantity: import_bcs.bcs.u64(),
    type_name: type_name.TypeName
  }
});
const FlashLoanBorrowed = new import_utils.MoveStruct({
  name: `${$moduleName}::FlashLoanBorrowed`,
  fields: {
    pool_id: import_bcs.bcs.Address,
    borrow_quantity: import_bcs.bcs.u64(),
    type_name: type_name.TypeName
  }
});
