// @mysten/deepbook-v3@0.26.6/dist/esm/utils/config.js
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __accessCheck = (obj, member, msg) => member.has(obj) || __typeError("Cannot " + msg);
var __privateGet = (obj, member, getter) => (__accessCheck(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd = (obj, member, value) => member.has(obj) ? __typeError("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet = (obj, member, value, setter) => (__accessCheck(obj, member, "write to private field"), setter ? setter.call(obj, value) : member.set(obj, value), value);
var _coins, _pools, _marginPools;
import { normalizeSuiAddress } from "@mysten/sui/utils";
import { BalanceManagerContract } from "../transactions/balanceManager.js";
import { ResourceNotFoundError, ErrorMessages } from "./errors.js";
import {
  mainnetCoins,
  mainnetPackageIds,
  mainnetPools,
  testnetCoins,
  testnetPackageIds,
  testnetPools,
  mainnetMarginPools,
  testnetMarginPools,
  mainnetPythConfigs,
  testnetPythConfigs
} from "./constants.js";
const FLOAT_SCALAR = 1e9;
const DEEP_SCALAR = 1e6;
const MAX_TIMESTAMP = 1844674407370955161n;
const PRICE_INFO_OBJECT_MAX_AGE_MS = 5e3;
const GAS_BUDGET = 25e7;
const POOL_CREATION_FEE_DEEP = 5e8;
class DeepBookConfig {
  constructor({
    env,
    address,
    adminCap,
    marginAdminCap,
    marginMaintainerCap,
    balanceManagers,
    marginManagers,
    coins,
    pools,
    marginPools
  }) {
    __privateAdd(this, _coins);
    __privateAdd(this, _pools);
    __privateAdd(this, _marginPools);
    this.env = env;
    this.address = normalizeSuiAddress(address);
    this.adminCap = adminCap;
    this.marginAdminCap = marginAdminCap;
    this.marginMaintainerCap = marginMaintainerCap;
    this.balanceManagers = balanceManagers || {};
    this.marginManagers = marginManagers || {};
    if (env === "mainnet") {
      __privateSet(this, _coins, coins || mainnetCoins);
      __privateSet(this, _pools, pools || mainnetPools);
      __privateSet(this, _marginPools, marginPools || mainnetMarginPools);
      this.DEEPBOOK_PACKAGE_ID = mainnetPackageIds.DEEPBOOK_PACKAGE_ID;
      this.REGISTRY_ID = mainnetPackageIds.REGISTRY_ID;
      this.DEEP_TREASURY_ID = mainnetPackageIds.DEEP_TREASURY_ID;
      this.MARGIN_PACKAGE_ID = mainnetPackageIds.MARGIN_PACKAGE_ID;
      this.MARGIN_REGISTRY_ID = mainnetPackageIds.MARGIN_REGISTRY_ID;
      this.LIQUIDATION_PACKAGE_ID = mainnetPackageIds.LIQUIDATION_PACKAGE_ID;
      this.pyth = mainnetPythConfigs;
    } else {
      __privateSet(this, _coins, coins || testnetCoins);
      __privateSet(this, _pools, pools || testnetPools);
      __privateSet(this, _marginPools, marginPools || testnetMarginPools);
      this.DEEPBOOK_PACKAGE_ID = testnetPackageIds.DEEPBOOK_PACKAGE_ID;
      this.REGISTRY_ID = testnetPackageIds.REGISTRY_ID;
      this.DEEP_TREASURY_ID = testnetPackageIds.DEEP_TREASURY_ID;
      this.MARGIN_PACKAGE_ID = testnetPackageIds.MARGIN_PACKAGE_ID;
      this.MARGIN_REGISTRY_ID = testnetPackageIds.MARGIN_REGISTRY_ID;
      this.LIQUIDATION_PACKAGE_ID = testnetPackageIds.LIQUIDATION_PACKAGE_ID;
      this.pyth = testnetPythConfigs;
    }
    this.balanceManager = new BalanceManagerContract(this);
  }
  // Getters
  getCoin(key) {
    const coin = __privateGet(this, _coins)[key];
    if (!coin) {
      throw new ResourceNotFoundError("Coin", key);
    }
    return coin;
  }
  getPool(key) {
    const pool = __privateGet(this, _pools)[key];
    if (!pool) {
      throw new ResourceNotFoundError("Pool", key);
    }
    return pool;
  }
  getMarginPool(key) {
    const pool = __privateGet(this, _marginPools)[key];
    if (!pool) {
      throw new ResourceNotFoundError("Margin pool", key);
    }
    return pool;
  }
  /**
   * @description Get the balance manager by key
   * @param managerKey Key of the balance manager
   * @returns The BalanceManager object
   */
  getBalanceManager(managerKey) {
    if (!Object.hasOwn(this.balanceManagers, managerKey)) {
      throw new Error(ErrorMessages.BALANCE_MANAGER_NOT_FOUND(managerKey));
    }
    return this.balanceManagers[managerKey];
  }
  /**
   * @description Get the margin manager by key
   * @param managerKey Key of the margin manager
   * @returns The MarginManager object
   */
  getMarginManager(managerKey) {
    if (!Object.hasOwn(this.marginManagers, managerKey)) {
      throw new Error(ErrorMessages.MARGIN_MANAGER_NOT_FOUND(managerKey));
    }
    return this.marginManagers[managerKey];
  }
}
_coins = new WeakMap();
_pools = new WeakMap();
_marginPools = new WeakMap();
export {
  DEEP_SCALAR,
  DeepBookConfig,
  FLOAT_SCALAR,
  GAS_BUDGET,
  MAX_TIMESTAMP,
  POOL_CREATION_FEE_DEEP,
  PRICE_INFO_OBJECT_MAX_AGE_MS
};
