// @mysten/deepbook-v3@0.26.6/dist/esm/utils/errors.js
class DeepBookError extends Error {
  constructor(message) {
    super(message);
    this.name = "DeepBookError";
  }
}
class ResourceNotFoundError extends DeepBookError {
  constructor(resourceType, key) {
    super(`${resourceType} not found for key: ${key}`);
    this.name = "ResourceNotFoundError";
  }
}
class ConfigurationError extends DeepBookError {
  constructor(message) {
    super(message);
    this.name = "ConfigurationError";
  }
}
class ValidationError extends DeepBookError {
  constructor(message) {
    super(message);
    this.name = "ValidationError";
  }
}
const ErrorMessages = {
  // Configuration errors
  ADMIN_CAP_NOT_SET: "Admin capability not configured",
  MARGIN_ADMIN_CAP_NOT_SET: "Margin admin capability not configured",
  MARGIN_MAINTAINER_CAP_NOT_SET: "Margin maintainer capability not configured",
  // Resource errors
  COIN_NOT_FOUND: (key) => `Coin not found for key: ${key}`,
  POOL_NOT_FOUND: (key) => `Pool not found for key: ${key}`,
  MARGIN_POOL_NOT_FOUND: (key) => `Margin pool not found for key: ${key}`,
  BALANCE_MANAGER_NOT_FOUND: (key) => `Balance manager with key ${key} not found`,
  MARGIN_MANAGER_NOT_FOUND: (key) => `Margin manager with key ${key} not found`,
  PRICE_INFO_NOT_FOUND: (coinKey) => `Price info object not found for ${coinKey}`,
  // Validation errors
  INVALID_ARGUMENT_COUNT: (expected, got) => `Invalid number of arguments, expected ${expected}, got ${got}`,
  PARAMETER_REQUIRED: (name) => `Parameter ${name} is required`,
  INVALID_ARGUMENT: (arg, type) => `Invalid argument ${arg} for type ${type}`,
  INVALID_ADDRESS: "Address must be a valid Sui address"
};
export {
  ConfigurationError,
  DeepBookError,
  ErrorMessages,
  ResourceNotFoundError,
  ValidationError
};
