// @mysten/deepbook-v3@0.26.6/dist/esm/utils/validation.js
import { isValidSuiAddress } from "@mysten/sui/utils";
import { ValidationError, ConfigurationError } from "./errors.js";
function validateRequired(value, errorMessage) {
  if (value === void 0 || value === null) {
    throw new ConfigurationError(errorMessage);
  }
  return value;
}
function validateAddress(address, fieldName = "Address") {
  if (!isValidSuiAddress(address)) {
    throw new ValidationError(`${fieldName} must be a valid Sui address`);
  }
  return address;
}
function validatePositiveNumber(value, fieldName) {
  if (value <= 0) {
    throw new ValidationError(`${fieldName} must be a positive number`);
  }
  return value;
}
function validateNonNegativeNumber(value, fieldName) {
  if (value < 0) {
    throw new ValidationError(`${fieldName} must be non-negative`);
  }
  return value;
}
function validateRange(value, min, max, fieldName) {
  if (value < min || value > max) {
    throw new ValidationError(`${fieldName} must be between ${min} and ${max}`);
  }
  return value;
}
function validateNonEmptyArray(array, fieldName) {
  if (!array || array.length === 0) {
    throw new ValidationError(`${fieldName} cannot be empty`);
  }
  return array;
}
export {
  validateAddress,
  validateNonEmptyArray,
  validateNonNegativeNumber,
  validatePositiveNumber,
  validateRange,
  validateRequired
};
