// @mysten/deepbook-v3@0.26.6/dist/cjs/utils/errors.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var errors_exports = {};
__export(errors_exports, {
  ConfigurationError: () => ConfigurationError,
  DeepBookError: () => DeepBookError,
  ErrorMessages: () => ErrorMessages,
  ResourceNotFoundError: () => ResourceNotFoundError,
  ValidationError: () => ValidationError
});
module.exports = __toCommonJS(errors_exports);
class DeepBookError extends Error {
  constructor(message) {
    super(message);
    this.name = "DeepBookError";
  }
}
class ResourceNotFoundError extends DeepBookError {
  constructor(resourceType, key) {
    super(`${resourceType} not found for key: ${key}`);
    this.name = "ResourceNotFoundError";
  }
}
class ConfigurationError extends DeepBookError {
  constructor(message) {
    super(message);
    this.name = "ConfigurationError";
  }
}
class ValidationError extends DeepBookError {
  constructor(message) {
    super(message);
    this.name = "ValidationError";
  }
}
const ErrorMessages = {
  // Configuration errors
  ADMIN_CAP_NOT_SET: "Admin capability not configured",
  MARGIN_ADMIN_CAP_NOT_SET: "Margin admin capability not configured",
  MARGIN_MAINTAINER_CAP_NOT_SET: "Margin maintainer capability not configured",
  // Resource errors
  COIN_NOT_FOUND: (key) => `Coin not found for key: ${key}`,
  POOL_NOT_FOUND: (key) => `Pool not found for key: ${key}`,
  MARGIN_POOL_NOT_FOUND: (key) => `Margin pool not found for key: ${key}`,
  BALANCE_MANAGER_NOT_FOUND: (key) => `Balance manager with key ${key} not found`,
  MARGIN_MANAGER_NOT_FOUND: (key) => `Margin manager with key ${key} not found`,
  PRICE_INFO_NOT_FOUND: (coinKey) => `Price info object not found for ${coinKey}`,
  // Validation errors
  INVALID_ARGUMENT_COUNT: (expected, got) => `Invalid number of arguments, expected ${expected}, got ${got}`,
  PARAMETER_REQUIRED: (name) => `Parameter ${name} is required`,
  INVALID_ARGUMENT: (arg, type) => `Invalid argument ${arg} for type ${type}`,
  INVALID_ADDRESS: "Address must be a valid Sui address"
};
