// @mysten/deepbook-v3@0.26.6/dist/cjs/utils/validation.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var validation_exports = {};
__export(validation_exports, {
  validateAddress: () => validateAddress,
  validateNonEmptyArray: () => validateNonEmptyArray,
  validateNonNegativeNumber: () => validateNonNegativeNumber,
  validatePositiveNumber: () => validatePositiveNumber,
  validateRange: () => validateRange,
  validateRequired: () => validateRequired
});
module.exports = __toCommonJS(validation_exports);
var import_utils = require("@mysten/sui/utils");
var import_errors = require("./errors.js");
function validateRequired(value, errorMessage) {
  if (value === void 0 || value === null) {
    throw new import_errors.ConfigurationError(errorMessage);
  }
  return value;
}
function validateAddress(address, fieldName = "Address") {
  if (!(0, import_utils.isValidSuiAddress)(address)) {
    throw new import_errors.ValidationError(`${fieldName} must be a valid Sui address`);
  }
  return address;
}
function validatePositiveNumber(value, fieldName) {
  if (value <= 0) {
    throw new import_errors.ValidationError(`${fieldName} must be a positive number`);
  }
  return value;
}
function validateNonNegativeNumber(value, fieldName) {
  if (value < 0) {
    throw new import_errors.ValidationError(`${fieldName} must be non-negative`);
  }
  return value;
}
function validateRange(value, min, max, fieldName) {
  if (value < min || value > max) {
    throw new import_errors.ValidationError(`${fieldName} must be between ${min} and ${max}`);
  }
  return value;
}
function validateNonEmptyArray(array, fieldName) {
  if (!array || array.length === 0) {
    throw new import_errors.ValidationError(`${fieldName} cannot be empty`);
  }
  return array;
}
