// @mysten/deepbook-v3@0.26.6/dist/esm/transactions/deepbookAdmin.js
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __accessCheck = (obj, member, msg) => member.has(obj) || __typeError("Cannot " + msg);
var __privateGet = (obj, member, getter) => (__accessCheck(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd = (obj, member, value) => member.has(obj) ? __typeError("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet = (obj, member, value, setter) => (__accessCheck(obj, member, "write to private field"), setter ? setter.call(obj, value) : member.set(obj, value), value);
var __privateMethod = (obj, member, method) => (__accessCheck(obj, member, "access private method"), method);
var _config, _DeepBookAdminContract_instances, adminCap_fn;
import { FLOAT_SCALAR } from "../utils/config.js";
class DeepBookAdminContract {
  /**
   * @param {DeepBookConfig} config Configuration for DeepBookAdminContract
   */
  constructor(config) {
    __privateAdd(this, _DeepBookAdminContract_instances);
    __privateAdd(this, _config);
    /**
     * @description Create a new pool as admin
     * @param {CreatePoolAdminParams} params Parameters for creating pool as admin
     * @returns A function that takes a Transaction object
     */
    this.createPoolAdmin = (params) => (tx) => {
      tx.setSenderIfNotSet(__privateGet(this, _config).address);
      const { baseCoinKey, quoteCoinKey, tickSize, lotSize, minSize, whitelisted, stablePool } = params;
      const baseCoin = __privateGet(this, _config).getCoin(baseCoinKey);
      const quoteCoin = __privateGet(this, _config).getCoin(quoteCoinKey);
      const baseScalar = baseCoin.scalar;
      const quoteScalar = quoteCoin.scalar;
      const adjustedTickSize = Math.round(tickSize * FLOAT_SCALAR * quoteScalar / baseScalar);
      const adjustedLotSize = Math.round(lotSize * baseScalar);
      const adjustedMinSize = Math.round(minSize * baseScalar);
      tx.moveCall({
        target: `${__privateGet(this, _config).DEEPBOOK_PACKAGE_ID}::pool::create_pool_admin`,
        arguments: [
          tx.object(__privateGet(this, _config).REGISTRY_ID),
          // registry_id
          tx.pure.u64(adjustedTickSize),
          // adjusted tick_size
          tx.pure.u64(adjustedLotSize),
          // adjusted lot_size
          tx.pure.u64(adjustedMinSize),
          // adjusted min_size
          tx.pure.bool(whitelisted),
          tx.pure.bool(stablePool),
          tx.object(__privateMethod(this, _DeepBookAdminContract_instances, adminCap_fn).call(this))
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Unregister a pool as admin
     * @param {string} poolKey The key of the pool to be unregistered by admin
     * @returns A function that takes a Transaction object
     */
    this.unregisterPoolAdmin = (poolKey) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      tx.moveCall({
        target: `${__privateGet(this, _config).DEEPBOOK_PACKAGE_ID}::pool::unregister_pool_admin`,
        arguments: [
          tx.object(pool.address),
          tx.object(__privateGet(this, _config).REGISTRY_ID),
          tx.object(__privateMethod(this, _DeepBookAdminContract_instances, adminCap_fn).call(this))
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Update the allowed versions for a pool
     * @param {string} poolKey The key of the pool to be updated
     * @returns A function that takes a Transaction object
     */
    this.updateAllowedVersions = (poolKey) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      tx.moveCall({
        target: `${__privateGet(this, _config).DEEPBOOK_PACKAGE_ID}::pool::update_allowed_versions`,
        arguments: [
          tx.object(pool.address),
          tx.object(__privateGet(this, _config).REGISTRY_ID),
          tx.object(__privateMethod(this, _DeepBookAdminContract_instances, adminCap_fn).call(this))
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Enable a specific version
     * @param {number} version The version to be enabled
     * @returns A function that takes a Transaction object
     */
    this.enableVersion = (version) => (tx) => {
      tx.moveCall({
        target: `${__privateGet(this, _config).DEEPBOOK_PACKAGE_ID}::registry::enable_version`,
        arguments: [
          tx.object(__privateGet(this, _config).REGISTRY_ID),
          tx.pure.u64(version),
          tx.object(__privateMethod(this, _DeepBookAdminContract_instances, adminCap_fn).call(this))
        ]
      });
    };
    /**
     * @description Disable a specific version
     * @param {number} version The version to be disabled
     * @returns A function that takes a Transaction object
     */
    this.disableVersion = (version) => (tx) => {
      tx.moveCall({
        target: `${__privateGet(this, _config).DEEPBOOK_PACKAGE_ID}::registry::disable_version`,
        arguments: [
          tx.object(__privateGet(this, _config).REGISTRY_ID),
          tx.pure.u64(version),
          tx.object(__privateMethod(this, _DeepBookAdminContract_instances, adminCap_fn).call(this))
        ]
      });
    };
    /**
     * @description Sets the treasury address where pool creation fees will be sent
     * @param {string} treasuryAddress The treasury address
     * @returns A function that takes a Transaction object
     */
    this.setTreasuryAddress = (treasuryAddress) => (tx) => {
      tx.moveCall({
        target: `${__privateGet(this, _config).DEEPBOOK_PACKAGE_ID}::registry::set_treasury_address`,
        arguments: [
          tx.object(__privateGet(this, _config).REGISTRY_ID),
          tx.pure.address(treasuryAddress),
          tx.object(__privateMethod(this, _DeepBookAdminContract_instances, adminCap_fn).call(this))
        ]
      });
    };
    /**
     * @description Add a coin to whitelist of stable coins
     * @param {string} stableCoinKey The name of the stable coin to be added
     * @returns A function that takes a Transaction object
     */
    this.addStableCoin = (stableCoinKey) => (tx) => {
      const stableCoinType = __privateGet(this, _config).getCoin(stableCoinKey).type;
      tx.moveCall({
        target: `${__privateGet(this, _config).DEEPBOOK_PACKAGE_ID}::registry::add_stablecoin`,
        arguments: [tx.object(__privateGet(this, _config).REGISTRY_ID), tx.object(__privateMethod(this, _DeepBookAdminContract_instances, adminCap_fn).call(this))],
        typeArguments: [stableCoinType]
      });
    };
    /**
     * @description Remove a coin from whitelist of stable coins
     * @param {string} stableCoinKey The name of the stable coin to be removed
     * @returns A function that takes a Transaction object
     */
    this.removeStableCoin = (stableCoinKey) => (tx) => {
      const stableCoinType = __privateGet(this, _config).getCoin(stableCoinKey).type;
      tx.moveCall({
        target: `${__privateGet(this, _config).DEEPBOOK_PACKAGE_ID}::registry::remove_stablecoin`,
        arguments: [tx.object(__privateGet(this, _config).REGISTRY_ID), tx.object(__privateMethod(this, _DeepBookAdminContract_instances, adminCap_fn).call(this))],
        typeArguments: [stableCoinType]
      });
    };
    /**
     * @description Adjust the tick size of a pool
     * @param {string} poolKey The key to identify the pool
     * @param {number} newTickSize The new tick size
     * @returns A function that takes a Transaction object
     */
    this.adjustTickSize = (poolKey, newTickSize) => (tx) => {
      tx.setSenderIfNotSet(__privateGet(this, _config).address);
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const baseScalar = baseCoin.scalar;
      const quoteScalar = quoteCoin.scalar;
      const adjustedTickSize = newTickSize * FLOAT_SCALAR * quoteScalar / baseScalar;
      tx.moveCall({
        target: `${__privateGet(this, _config).DEEPBOOK_PACKAGE_ID}::pool::adjust_tick_size_admin`,
        arguments: [
          tx.object(pool.address),
          // pool address
          tx.pure.u64(adjustedTickSize),
          // adjusted tick_size
          tx.object(__privateMethod(this, _DeepBookAdminContract_instances, adminCap_fn).call(this)),
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Adjust the lot size and min size of a pool
     * @param {string} poolKey The key to identify the pool
     * @param {number} newLotSize The new lot size
     * @param {number} newMinSize The new min size
     * @returns A function that takes a Transaction object
     */
    this.adjustMinLotSize = (poolKey, newLotSize, newMinSize) => (tx) => {
      tx.setSenderIfNotSet(__privateGet(this, _config).address);
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const baseScalar = baseCoin.scalar;
      const adjustedLotSize = newLotSize * baseScalar;
      const adjustedMinSize = newMinSize * baseScalar;
      tx.moveCall({
        target: `${__privateGet(this, _config).DEEPBOOK_PACKAGE_ID}::pool::adjust_min_lot_size_admin`,
        arguments: [
          tx.object(pool.address),
          // pool address
          tx.pure.u64(adjustedLotSize),
          tx.pure.u64(adjustedMinSize),
          tx.object(__privateMethod(this, _DeepBookAdminContract_instances, adminCap_fn).call(this)),
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Initialize the balance manager map
     * @returns A function that takes a Transaction object
     */
    this.initBalanceManagerMap = () => (tx) => {
      tx.moveCall({
        target: `${__privateGet(this, _config).DEEPBOOK_PACKAGE_ID}::registry::init_balance_manager_map`,
        arguments: [tx.object(__privateGet(this, _config).REGISTRY_ID), tx.object(__privateMethod(this, _DeepBookAdminContract_instances, adminCap_fn).call(this))]
      });
    };
    /**
     * @description Set the EWMA parameters for a pool
     * @param {string} poolKey The key to identify the pool
     * @param {SetEwmaParamsParams} params The parameters to set
     * @returns A function that takes a Transaction object
     */
    this.setEwmaParams = (poolKey, params) => (tx) => {
      const { alpha, zScoreThreshold, additionalTakerFee } = params;
      const adjustedAlpha = Math.round(alpha * FLOAT_SCALAR);
      const adjustedZScoreThreshold = Math.round(zScoreThreshold * FLOAT_SCALAR);
      const adjustedAdditionalTakerFee = Math.round(additionalTakerFee * FLOAT_SCALAR);
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      tx.moveCall({
        target: `${__privateGet(this, _config).DEEPBOOK_PACKAGE_ID}::pool::set_ewma_params`,
        arguments: [
          tx.object(pool.address),
          tx.object(__privateMethod(this, _DeepBookAdminContract_instances, adminCap_fn).call(this)),
          tx.pure.u64(adjustedAlpha),
          tx.pure.u64(adjustedZScoreThreshold),
          tx.pure.u64(adjustedAdditionalTakerFee),
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Enable or disable the EWMA state for a pool
     * @param {string} poolKey The key to identify the pool
     * @param {boolean} enable Whether to enable or disable the EWMA state
     * @returns A function that takes a Transaction object
     */
    this.enableEwmaState = (poolKey, enable) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      tx.moveCall({
        target: `${__privateGet(this, _config).DEEPBOOK_PACKAGE_ID}::pool::enable_ewma_state`,
        arguments: [
          tx.object(pool.address),
          tx.object(__privateMethod(this, _DeepBookAdminContract_instances, adminCap_fn).call(this)),
          tx.pure.bool(enable),
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Authorize the MarginApp to access protected features of DeepBook
     * @returns A function that takes a Transaction object
     */
    this.authorizeMarginApp = () => (tx) => {
      tx.moveCall({
        target: `${__privateGet(this, _config).DEEPBOOK_PACKAGE_ID}::registry::authorize_app`,
        arguments: [tx.object(__privateGet(this, _config).REGISTRY_ID), tx.object(__privateMethod(this, _DeepBookAdminContract_instances, adminCap_fn).call(this))],
        typeArguments: [`${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::MarginApp`]
      });
    };
    /**
     * @description Deauthorize the MarginApp by removing its authorization key
     * @returns A function that takes a Transaction object and returns a bool
     */
    this.deauthorizeMarginApp = () => (tx) => {
      return tx.moveCall({
        target: `${__privateGet(this, _config).DEEPBOOK_PACKAGE_ID}::registry::deauthorize_app`,
        arguments: [tx.object(__privateGet(this, _config).REGISTRY_ID), tx.object(__privateMethod(this, _DeepBookAdminContract_instances, adminCap_fn).call(this))],
        typeArguments: [`${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::MarginApp`]
      });
    };
    __privateSet(this, _config, config);
  }
}
_config = new WeakMap();
_DeepBookAdminContract_instances = new WeakSet();
/**
 * @returns The admin capability required for admin operations
 * @throws Error if the admin capability is not set
 */
adminCap_fn = function() {
  const adminCap = __privateGet(this, _config).adminCap;
  if (!adminCap) {
    throw new Error("ADMIN_CAP environment variable not set");
  }
  return adminCap;
};
export {
  DeepBookAdminContract
};
