// @mysten/deepbook-v3@0.26.6/dist/esm/transactions/marginPool.js
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __accessCheck = (obj, member, msg) => member.has(obj) || __typeError("Cannot " + msg);
var __privateGet = (obj, member, getter) => (__accessCheck(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd = (obj, member, value) => member.has(obj) ? __typeError("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet = (obj, member, value, setter) => (__accessCheck(obj, member, "write to private field"), setter ? setter.call(obj, value) : member.set(obj, value), value);
var _config;
import { coinWithBalance } from "@mysten/sui/transactions";
class MarginPoolContract {
  /**
   * @param {DeepBookConfig} config Configuration for MarginPoolContract
   */
  constructor(config) {
    __privateAdd(this, _config);
    /**
     * @description Mint a supplier cap for margin pool
     * @returns A function that takes a Transaction object
     */
    this.mintSupplierCap = () => (tx) => {
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_pool::mint_supplier_cap`,
        arguments: [tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID), tx.object.clock()]
      });
    };
    /**
     * @description Supply to a margin pool
     * @param {string} coinKey The key to identify the pool
     * @param {TransactionObjectArgument} supplierCap The supplier cap object
     * @param {number} amountToDeposit The amount to deposit
     * @param {string} referralId The ID of the referral
     * @returns A function that takes a Transaction object
     */
    this.supplyToMarginPool = (coinKey, supplierCap, amountToDeposit, referralId) => (tx) => {
      tx.setSenderIfNotSet(__privateGet(this, _config).address);
      const marginPool = __privateGet(this, _config).getMarginPool(coinKey);
      const coin = __privateGet(this, _config).getCoin(coinKey);
      const depositInput = Math.round(amountToDeposit * coin.scalar);
      const supply = coinWithBalance({
        type: coin.type,
        balance: depositInput
      });
      tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_pool::supply`,
        arguments: [
          tx.object(marginPool.address),
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          supplierCap,
          supply,
          tx.object.option({
            type: "0x2::object::ID",
            value: referralId ? tx.pure.id(referralId) : null
          }),
          tx.object.clock()
        ],
        typeArguments: [marginPool.type]
      });
    };
    /**
     * @description Withdraw from a margin pool
     * @param {string} coinKey The key to identify the pool
     * @param {TransactionObjectArgument} supplierCap The supplier cap object
     * @param {number} amountToWithdraw The amount to withdraw
     * @returns A function that takes a Transaction object
     */
    this.withdrawFromMarginPool = (coinKey, supplierCap, amountToWithdraw) => (tx) => {
      const marginPool = __privateGet(this, _config).getMarginPool(coinKey);
      const coin = __privateGet(this, _config).getCoin(coinKey);
      const withdrawInput = amountToWithdraw ? tx.pure.u64(Math.round(amountToWithdraw * coin.scalar)) : null;
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_pool::withdraw`,
        arguments: [
          tx.object(marginPool.address),
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          supplierCap,
          tx.object.option({ type: "u64", value: withdrawInput }),
          tx.object.clock()
        ],
        typeArguments: [marginPool.type]
      });
    };
    /**
     * @description Mint a referral for a margin pool
     * @param {string} coinKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.mintSupplyReferral = (coinKey) => (tx) => {
      const marginPool = __privateGet(this, _config).getMarginPool(coinKey);
      tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_pool::mint_supply_referral`,
        arguments: [
          tx.object(marginPool.address),
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object.clock()
        ],
        typeArguments: [marginPool.type]
      });
    };
    /**
     * @description Withdraw referral fees from a margin pool
     * @param {string} coinKey The key to identify the pool
     * @param {string} referralId The ID of the referral
     * @returns A function that takes a Transaction object
     */
    this.withdrawReferralFees = (coinKey, referralId) => (tx) => {
      const marginPool = __privateGet(this, _config).getMarginPool(coinKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_pool::withdraw_referral_fees`,
        arguments: [
          tx.object(marginPool.address),
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(referralId)
        ],
        typeArguments: [marginPool.type]
      });
    };
    // === Read-only/View Functions ===
    /**
     * @description Get the margin pool ID
     * @param {string} coinKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.getId = (coinKey) => (tx) => {
      const marginPool = __privateGet(this, _config).getMarginPool(coinKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_pool::id`,
        arguments: [tx.object(marginPool.address)],
        typeArguments: [marginPool.type]
      });
    };
    /**
     * @description Check if a deepbook pool is allowed for borrowing
     * @param {string} coinKey The key to identify the margin pool
     * @param {string} deepbookPoolId The ID of the deepbook pool
     * @returns A function that takes a Transaction object
     */
    this.deepbookPoolAllowed = (coinKey, deepbookPoolId) => (tx) => {
      const marginPool = __privateGet(this, _config).getMarginPool(coinKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_pool::deepbook_pool_allowed`,
        arguments: [tx.object(marginPool.address), tx.pure.id(deepbookPoolId)],
        typeArguments: [marginPool.type]
      });
    };
    /**
     * @description Get the total supply amount
     * @param {string} coinKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.totalSupply = (coinKey) => (tx) => {
      const marginPool = __privateGet(this, _config).getMarginPool(coinKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_pool::total_supply`,
        arguments: [tx.object(marginPool.address)],
        typeArguments: [marginPool.type]
      });
    };
    /**
     * @description Get the total supply shares
     * @param {string} coinKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.supplyShares = (coinKey) => (tx) => {
      const marginPool = __privateGet(this, _config).getMarginPool(coinKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_pool::supply_shares`,
        arguments: [tx.object(marginPool.address)],
        typeArguments: [marginPool.type]
      });
    };
    /**
     * @description Get the total borrow amount
     * @param {string} coinKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.totalBorrow = (coinKey) => (tx) => {
      const marginPool = __privateGet(this, _config).getMarginPool(coinKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_pool::total_borrow`,
        arguments: [tx.object(marginPool.address)],
        typeArguments: [marginPool.type]
      });
    };
    /**
     * @description Get the total borrow shares
     * @param {string} coinKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.borrowShares = (coinKey) => (tx) => {
      const marginPool = __privateGet(this, _config).getMarginPool(coinKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_pool::borrow_shares`,
        arguments: [tx.object(marginPool.address)],
        typeArguments: [marginPool.type]
      });
    };
    /**
     * @description Get the last update timestamp
     * @param {string} coinKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.lastUpdateTimestamp = (coinKey) => (tx) => {
      const marginPool = __privateGet(this, _config).getMarginPool(coinKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_pool::last_update_timestamp`,
        arguments: [tx.object(marginPool.address)],
        typeArguments: [marginPool.type]
      });
    };
    /**
     * @description Get the supply cap
     * @param {string} coinKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.supplyCap = (coinKey) => (tx) => {
      const marginPool = __privateGet(this, _config).getMarginPool(coinKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_pool::supply_cap`,
        arguments: [tx.object(marginPool.address)],
        typeArguments: [marginPool.type]
      });
    };
    /**
     * @description Get the max utilization rate
     * @param {string} coinKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.maxUtilizationRate = (coinKey) => (tx) => {
      const marginPool = __privateGet(this, _config).getMarginPool(coinKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_pool::max_utilization_rate`,
        arguments: [tx.object(marginPool.address)],
        typeArguments: [marginPool.type]
      });
    };
    /**
     * @description Get the protocol spread
     * @param {string} coinKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.protocolSpread = (coinKey) => (tx) => {
      const marginPool = __privateGet(this, _config).getMarginPool(coinKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_pool::protocol_spread`,
        arguments: [tx.object(marginPool.address)],
        typeArguments: [marginPool.type]
      });
    };
    /**
     * @description Get the minimum borrow amount
     * @param {string} coinKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.minBorrow = (coinKey) => (tx) => {
      const marginPool = __privateGet(this, _config).getMarginPool(coinKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_pool::min_borrow`,
        arguments: [tx.object(marginPool.address)],
        typeArguments: [marginPool.type]
      });
    };
    /**
     * @description Get the current interest rate
     * @param {string} coinKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.interestRate = (coinKey) => (tx) => {
      const marginPool = __privateGet(this, _config).getMarginPool(coinKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_pool::interest_rate`,
        arguments: [tx.object(marginPool.address)],
        typeArguments: [marginPool.type]
      });
    };
    /**
     * @description Get user supply shares for a supplier cap
     * @param {string} coinKey The key to identify the pool
     * @param {string} supplierCapId The ID of the supplier cap
     * @returns A function that takes a Transaction object
     */
    this.userSupplyShares = (coinKey, supplierCapId) => (tx) => {
      const marginPool = __privateGet(this, _config).getMarginPool(coinKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_pool::user_supply_shares`,
        arguments: [tx.object(marginPool.address), tx.pure.id(supplierCapId)],
        typeArguments: [marginPool.type]
      });
    };
    /**
     * @description Get user supply amount for a supplier cap
     * @param {string} coinKey The key to identify the pool
     * @param {string} supplierCapId The ID of the supplier cap
     * @returns A function that takes a Transaction object
     */
    this.userSupplyAmount = (coinKey, supplierCapId) => (tx) => {
      const marginPool = __privateGet(this, _config).getMarginPool(coinKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_pool::user_supply_amount`,
        arguments: [tx.object(marginPool.address), tx.pure.id(supplierCapId), tx.object.clock()],
        typeArguments: [marginPool.type]
      });
    };
    __privateSet(this, _config, config);
  }
}
_config = new WeakMap();
export {
  MarginPoolContract
};
