// @mysten/deepbook-v3@0.26.6/dist/esm/transactions/marginRegistry.js
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __accessCheck = (obj, member, msg) => member.has(obj) || __typeError("Cannot " + msg);
var __privateGet = (obj, member, getter) => (__accessCheck(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd = (obj, member, value) => member.has(obj) ? __typeError("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet = (obj, member, value, setter) => (__accessCheck(obj, member, "write to private field"), setter ? setter.call(obj, value) : member.set(obj, value), value);
var _config;
class MarginRegistryContract {
  /**
   * @param {DeepBookConfig} config Configuration for MarginRegistryContract
   */
  constructor(config) {
    __privateAdd(this, _config);
    /**
     * @description Check if a deepbook pool is enabled for margin trading
     * @param {string} poolKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.poolEnabled = (poolKey) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_registry::pool_enabled`,
        arguments: [tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID), tx.object(pool.address)],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Get the margin pool ID for a given asset
     * @param {string} coinKey The key to identify the coin
     * @returns A function that takes a Transaction object
     */
    this.getMarginPoolId = (coinKey) => (tx) => {
      const coin = __privateGet(this, _config).getCoin(coinKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_registry::get_margin_pool_id`,
        arguments: [tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID)],
        typeArguments: [coin.type]
      });
    };
    /**
     * @description Get the margin pool IDs (base and quote) for a deepbook pool
     * @param {string} poolKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.getDeepbookPoolMarginPoolIds = (poolKey) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_registry::get_deepbook_pool_margin_pool_ids`,
        arguments: [tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID), tx.pure.id(pool.address)],
        typeArguments: []
      });
    };
    /**
     * @description Get the margin manager IDs for a given owner
     * @param {string} owner The owner address
     * @returns A function that takes a Transaction object
     */
    this.getMarginManagerIds = (owner) => (tx) => {
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_registry::get_margin_manager_ids`,
        arguments: [tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID), tx.pure.address(owner)],
        typeArguments: []
      });
    };
    /**
     * @description Get the base margin pool ID for a deepbook pool
     * @param {string} poolKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.baseMarginPoolId = (poolKey) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_registry::base_margin_pool_id`,
        arguments: [tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID), tx.pure.id(pool.address)],
        typeArguments: []
      });
    };
    /**
     * @description Get the quote margin pool ID for a deepbook pool
     * @param {string} poolKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.quoteMarginPoolId = (poolKey) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_registry::quote_margin_pool_id`,
        arguments: [tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID), tx.pure.id(pool.address)],
        typeArguments: []
      });
    };
    /**
     * @description Get the minimum withdraw risk ratio for a deepbook pool
     * @param {string} poolKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.minWithdrawRiskRatio = (poolKey) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_registry::min_withdraw_risk_ratio`,
        arguments: [tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID), tx.pure.id(pool.address)],
        typeArguments: []
      });
    };
    /**
     * @description Get the minimum borrow risk ratio for a deepbook pool
     * @param {string} poolKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.minBorrowRiskRatio = (poolKey) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_registry::min_borrow_risk_ratio`,
        arguments: [tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID), tx.pure.id(pool.address)],
        typeArguments: []
      });
    };
    /**
     * @description Get the liquidation risk ratio for a deepbook pool
     * @param {string} poolKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.liquidationRiskRatio = (poolKey) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_registry::liquidation_risk_ratio`,
        arguments: [tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID), tx.pure.id(pool.address)],
        typeArguments: []
      });
    };
    /**
     * @description Get the target liquidation risk ratio for a deepbook pool
     * @param {string} poolKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.targetLiquidationRiskRatio = (poolKey) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_registry::target_liquidation_risk_ratio`,
        arguments: [tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID), tx.pure.id(pool.address)],
        typeArguments: []
      });
    };
    /**
     * @description Get the user liquidation reward for a deepbook pool
     * @param {string} poolKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.userLiquidationReward = (poolKey) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_registry::user_liquidation_reward`,
        arguments: [tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID), tx.pure.id(pool.address)],
        typeArguments: []
      });
    };
    /**
     * @description Get the pool liquidation reward for a deepbook pool
     * @param {string} poolKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.poolLiquidationReward = (poolKey) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_registry::pool_liquidation_reward`,
        arguments: [tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID), tx.pure.id(pool.address)],
        typeArguments: []
      });
    };
    /**
     * @description Get all allowed maintainer cap IDs
     * @returns A function that takes a Transaction object
     */
    this.allowedMaintainers = () => (tx) => {
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_registry::allowed_maintainers`,
        arguments: [tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID)],
        typeArguments: []
      });
    };
    /**
     * @description Get all allowed pause cap IDs
     * @returns A function that takes a Transaction object
     */
    this.allowedPauseCaps = () => (tx) => {
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_registry::allowed_pause_caps`,
        arguments: [tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID)],
        typeArguments: []
      });
    };
    __privateSet(this, _config, config);
  }
}
_config = new WeakMap();
export {
  MarginRegistryContract
};
