// @mysten/deepbook-v3@0.26.6/dist/esm/transactions/poolProxy.js
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __accessCheck = (obj, member, msg) => member.has(obj) || __typeError("Cannot " + msg);
var __privateGet = (obj, member, getter) => (__accessCheck(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd = (obj, member, value) => member.has(obj) ? __typeError("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet = (obj, member, value, setter) => (__accessCheck(obj, member, "write to private field"), setter ? setter.call(obj, value) : member.set(obj, value), value);
var _config;
import { OrderType, SelfMatchingOptions } from "../types/index.js";
import { MAX_TIMESTAMP, FLOAT_SCALAR } from "../utils/config.js";
class PoolProxyContract {
  /**
   * @param {DeepBookConfig} config Configuration for PoolProxyContract
   */
  constructor(config) {
    __privateAdd(this, _config);
    /**
     * @description Place a limit order
     * @param {PlaceMarginLimitOrderParams} params Parameters for placing a limit order
     * @returns A function that takes a Transaction object
     */
    this.placeLimitOrder = (params) => (tx) => {
      const {
        poolKey,
        marginManagerKey,
        clientOrderId,
        price,
        quantity,
        isBid,
        expiration = MAX_TIMESTAMP,
        orderType = OrderType.NO_RESTRICTION,
        selfMatchingOption = SelfMatchingOptions.SELF_MATCHING_ALLOWED,
        payWithDeep = true
      } = params;
      const pool = __privateGet(this, _config).getPool(poolKey);
      const manager = __privateGet(this, _config).getMarginManager(marginManagerKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const inputPrice = Math.round(price * FLOAT_SCALAR * quoteCoin.scalar / baseCoin.scalar);
      const inputQuantity = Math.round(quantity * baseCoin.scalar);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::pool_proxy::place_limit_order`,
        arguments: [
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(manager.address),
          tx.object(pool.address),
          tx.pure.u64(clientOrderId),
          tx.pure.u8(orderType),
          tx.pure.u8(selfMatchingOption),
          tx.pure.u64(inputPrice),
          tx.pure.u64(inputQuantity),
          tx.pure.bool(isBid),
          tx.pure.bool(payWithDeep),
          tx.pure.u64(expiration),
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Place a market order
     * @param {PlaceMarginMarketOrderParams} params Parameters for placing a market order
     * @returns A function that takes a Transaction object
     */
    this.placeMarketOrder = (params) => (tx) => {
      const {
        poolKey,
        marginManagerKey,
        clientOrderId,
        quantity,
        isBid,
        selfMatchingOption = SelfMatchingOptions.SELF_MATCHING_ALLOWED,
        payWithDeep = true
      } = params;
      const pool = __privateGet(this, _config).getPool(poolKey);
      const manager = __privateGet(this, _config).getMarginManager(marginManagerKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const inputQuantity = Math.round(quantity * baseCoin.scalar);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::pool_proxy::place_market_order`,
        arguments: [
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(manager.address),
          tx.object(pool.address),
          tx.pure.u64(clientOrderId),
          tx.pure.u8(selfMatchingOption),
          tx.pure.u64(inputQuantity),
          tx.pure.bool(isBid),
          tx.pure.bool(payWithDeep),
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Place a reduce only limit order
     * @param {PlaceMarginLimitOrderParams} params Parameters for placing a reduce only limit order
     * @returns A function that takes a Transaction object
     */
    this.placeReduceOnlyLimitOrder = (params) => (tx) => {
      const {
        poolKey,
        marginManagerKey,
        clientOrderId,
        price,
        quantity,
        isBid,
        expiration = MAX_TIMESTAMP,
        orderType = OrderType.NO_RESTRICTION,
        selfMatchingOption = SelfMatchingOptions.SELF_MATCHING_ALLOWED,
        payWithDeep = true
      } = params;
      const pool = __privateGet(this, _config).getPool(poolKey);
      const manager = __privateGet(this, _config).getMarginManager(marginManagerKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const inputPrice = Math.round(price * FLOAT_SCALAR * quoteCoin.scalar / baseCoin.scalar);
      const inputQuantity = Math.round(quantity * baseCoin.scalar);
      const marginPool = isBid ? __privateGet(this, _config).getMarginPool(pool.baseCoin) : __privateGet(this, _config).getMarginPool(pool.quoteCoin);
      const debtType = isBid ? baseCoin.type : quoteCoin.type;
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::pool_proxy::place_reduce_only_limit_order`,
        arguments: [
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(manager.address),
          tx.object(pool.address),
          tx.object(marginPool.address),
          tx.pure.u64(clientOrderId),
          tx.pure.u8(orderType),
          tx.pure.u8(selfMatchingOption),
          tx.pure.u64(inputPrice),
          tx.pure.u64(inputQuantity),
          tx.pure.bool(isBid),
          tx.pure.bool(payWithDeep),
          tx.pure.u64(expiration),
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type, debtType]
      });
    };
    /**
     * @description Place a reduce only market order
     * @param {PlaceMarginMarketOrderParams} params Parameters for placing a reduce only market order
     * @returns A function that takes a Transaction object
     */
    this.placeReduceOnlyMarketOrder = (params) => (tx) => {
      const {
        poolKey,
        marginManagerKey,
        clientOrderId,
        quantity,
        isBid,
        selfMatchingOption = SelfMatchingOptions.SELF_MATCHING_ALLOWED,
        payWithDeep = true
      } = params;
      const pool = __privateGet(this, _config).getPool(poolKey);
      const manager = __privateGet(this, _config).getMarginManager(marginManagerKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const inputQuantity = Math.round(quantity * baseCoin.scalar);
      const marginPool = isBid ? __privateGet(this, _config).getMarginPool(pool.baseCoin) : __privateGet(this, _config).getMarginPool(pool.quoteCoin);
      const debtType = isBid ? baseCoin.type : quoteCoin.type;
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::pool_proxy::place_reduce_only_market_order`,
        arguments: [
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(manager.address),
          tx.object(pool.address),
          tx.object(marginPool.address),
          tx.pure.u64(clientOrderId),
          tx.pure.u8(selfMatchingOption),
          tx.pure.u64(inputQuantity),
          tx.pure.bool(isBid),
          tx.pure.bool(payWithDeep),
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type, debtType]
      });
    };
    /**
     * @description Modify an existing order
     * @param {string} marginManagerKey The key to identify the MarginManager
     * @param {string} orderId Order ID to modify
     * @param {number} newQuantity New quantity for the order
     * @returns A function that takes a Transaction object
     */
    this.modifyOrder = (marginManagerKey, orderId, newQuantity) => (tx) => {
      const marginManager = __privateGet(this, _config).getMarginManager(marginManagerKey);
      const pool = __privateGet(this, _config).getPool(marginManager.poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const inputQuantity = Math.round(newQuantity * baseCoin.scalar);
      tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::pool_proxy::modify_order`,
        arguments: [
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(marginManager.address),
          tx.object(pool.address),
          tx.pure.u128(orderId),
          tx.pure.u64(inputQuantity),
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Cancel an existing order
     * @param {string} marginManagerKey The key to identify the MarginManager
     * @param {string} orderId Order ID to cancel
     * @returns A function that takes a Transaction object
     */
    this.cancelOrder = (marginManagerKey, orderId) => (tx) => {
      const marginManager = __privateGet(this, _config).getMarginManager(marginManagerKey);
      const pool = __privateGet(this, _config).getPool(marginManager.poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::pool_proxy::cancel_order`,
        arguments: [
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(marginManager.address),
          tx.object(pool.address),
          tx.pure.u128(orderId),
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Cancel multiple existing orders
     * @param {string} marginManagerKey The key to identify the MarginManager
     * @param {string[]} orderIds Order IDs to cancel
     * @returns A function that takes a Transaction object
     */
    this.cancelOrders = (marginManagerKey, orderIds) => (tx) => {
      const marginManager = __privateGet(this, _config).getMarginManager(marginManagerKey);
      const pool = __privateGet(this, _config).getPool(marginManager.poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::pool_proxy::cancel_orders`,
        arguments: [
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(marginManager.address),
          tx.object(pool.address),
          tx.pure.vector("u128", orderIds),
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Cancel all existing orders
     * @param {string} marginManagerKey The key to identify the MarginManager
     * @returns A function that takes a Transaction object
     */
    this.cancelAllOrders = (marginManagerKey) => (tx) => {
      const marginManager = __privateGet(this, _config).getMarginManager(marginManagerKey);
      const pool = __privateGet(this, _config).getPool(marginManager.poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::pool_proxy::cancel_all_orders`,
        arguments: [
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(marginManager.address),
          tx.object(pool.address),
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Withdraw settled amounts
     * @param {string} marginManagerKey The key to identify the MarginManager
     * @returns A function that takes a Transaction object
     */
    this.withdrawSettledAmounts = (marginManagerKey) => (tx) => {
      const marginManager = __privateGet(this, _config).getMarginManager(marginManagerKey);
      const pool = __privateGet(this, _config).getPool(marginManager.poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::pool_proxy::withdraw_settled_amounts`,
        arguments: [
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(marginManager.address),
          tx.object(pool.address)
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Stake in the pool
     * @param {string} marginManagerKey The key to identify the MarginManager
     * @param {number} stakeAmount The amount to stake
     * @returns A function that takes a Transaction object
     */
    this.stake = (marginManagerKey, stakeAmount) => (tx) => {
      const marginManager = __privateGet(this, _config).getMarginManager(marginManagerKey);
      const pool = __privateGet(this, _config).getPool(marginManager.poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const deepCoin = __privateGet(this, _config).getCoin("DEEP");
      const stakeInput = Math.round(stakeAmount * deepCoin.scalar);
      tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::pool_proxy::stake`,
        arguments: [
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(marginManager.address),
          tx.object(pool.address),
          tx.pure.u64(stakeInput)
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Unstake from the pool
     * @param {string} marginManagerKey The key to identify the MarginManager
     * @returns A function that takes a Transaction object
     */
    this.unstake = (marginManagerKey) => (tx) => {
      const marginManager = __privateGet(this, _config).getMarginManager(marginManagerKey);
      const pool = __privateGet(this, _config).getPool(marginManager.poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::pool_proxy::unstake`,
        arguments: [
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(marginManager.address),
          tx.object(pool.address)
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Submit a proposal
     * @param {string} marginManagerKey The key to identify the MarginManager
     * @param {MarginProposalParams} params Parameters for the proposal
     * @returns A function that takes a Transaction object
     */
    this.submitProposal = (marginManagerKey, params) => (tx) => {
      const { takerFee, makerFee, stakeRequired } = params;
      const marginManager = __privateGet(this, _config).getMarginManager(marginManagerKey);
      const pool = __privateGet(this, _config).getPool(marginManager.poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const stakeInput = Math.round(stakeRequired * FLOAT_SCALAR);
      const takerFeeInput = Math.round(takerFee * FLOAT_SCALAR);
      const makerFeeInput = Math.round(makerFee * FLOAT_SCALAR);
      tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::pool_proxy::submit_proposal`,
        arguments: [
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(marginManager.address),
          tx.object(pool.address),
          tx.pure.u64(takerFeeInput),
          tx.pure.u64(makerFeeInput),
          tx.pure.u64(stakeInput)
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Vote on a proposal
     * @param {string} marginManagerKey The key to identify the MarginManager
     * @param {string} proposalId The ID of the proposal to vote on
     * @returns A function that takes a Transaction object
     */
    this.vote = (marginManagerKey, proposalId) => (tx) => {
      const marginManager = __privateGet(this, _config).getMarginManager(marginManagerKey);
      const pool = __privateGet(this, _config).getPool(marginManager.poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::pool_proxy::vote`,
        arguments: [
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(marginManager.address),
          tx.object(pool.address),
          tx.pure.id(proposalId)
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Claim a rebate from a pool
     * @param {string} marginManagerKey The key to identify the MarginManager
     * @returns A function that takes a Transaction object
     */
    this.claimRebate = (marginManagerKey) => (tx) => {
      const marginManager = __privateGet(this, _config).getMarginManager(marginManagerKey);
      const pool = __privateGet(this, _config).getPool(marginManager.poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::pool_proxy::claim_rebate`,
        arguments: [
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(marginManager.address),
          tx.object(pool.address)
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    __privateSet(this, _config, config);
  }
}
_config = new WeakMap();
export {
  PoolProxyContract
};
