// @mysten/deepbook-v3@0.26.6/dist/esm/contracts/deepbook/balance_manager.js
import { MoveStruct, normalizeMoveArguments } from "../utils/index.js";
import { bcs } from "@mysten/sui/bcs";
import * as object from "./deps/sui/object.js";
import * as bag from "./deps/sui/bag.js";
import * as vec_set from "./deps/sui/vec_set.js";
import * as type_name from "./deps/std/type_name.js";
const $moduleName = "@deepbook/core::balance_manager";
const BalanceManager = new MoveStruct({
  name: `${$moduleName}::BalanceManager`,
  fields: {
    id: object.UID,
    owner: bcs.Address,
    balances: bag.Bag,
    allow_listed: vec_set.VecSet(bcs.Address)
  }
});
const BalanceManagerEvent = new MoveStruct({
  name: `${$moduleName}::BalanceManagerEvent`,
  fields: {
    balance_manager_id: bcs.Address,
    owner: bcs.Address
  }
});
const BalanceEvent = new MoveStruct({
  name: `${$moduleName}::BalanceEvent`,
  fields: {
    balance_manager_id: bcs.Address,
    asset: type_name.TypeName,
    amount: bcs.u64(),
    deposit: bcs.bool()
  }
});
const BalanceKey = new MoveStruct({
  name: `${$moduleName}::BalanceKey`,
  fields: {
    dummy_field: bcs.bool()
  }
});
const TradeCap = new MoveStruct({
  name: `${$moduleName}::TradeCap`,
  fields: {
    id: object.UID,
    balance_manager_id: bcs.Address
  }
});
const DepositCap = new MoveStruct({
  name: `${$moduleName}::DepositCap`,
  fields: {
    id: object.UID,
    balance_manager_id: bcs.Address
  }
});
const WithdrawCap = new MoveStruct({
  name: `${$moduleName}::WithdrawCap`,
  fields: {
    id: object.UID,
    balance_manager_id: bcs.Address
  }
});
const TradeProof = new MoveStruct({
  name: `${$moduleName}::TradeProof`,
  fields: {
    balance_manager_id: bcs.Address,
    trader: bcs.Address
  }
});
function _new(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "new"
  });
}
function newWithOwner(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["address"];
  const parameterNames = ["Owner"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "new_with_owner",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function newWithCustomOwner(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["address"];
  const parameterNames = ["owner"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "new_with_custom_owner",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function newWithCustomOwnerAndCaps(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["address"];
  const parameterNames = ["owner"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "new_with_custom_owner_and_caps",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function balance(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::balance_manager::BalanceManager`];
  const parameterNames = ["balanceManager"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "balance",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function mintTradeCap(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::balance_manager::BalanceManager`];
  const parameterNames = ["balanceManager"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "mint_trade_cap",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function mintDepositCap(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::balance_manager::BalanceManager`];
  const parameterNames = ["balanceManager"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "mint_deposit_cap",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function mintWithdrawCap(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::balance_manager::BalanceManager`];
  const parameterNames = ["balanceManager"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "mint_withdraw_cap",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function revokeTradeCap(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::balance_manager::BalanceManager`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID"
  ];
  const parameterNames = ["balanceManager", "tradeCapId"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "revoke_trade_cap",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function generateProofAsOwner(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::balance_manager::BalanceManager`];
  const parameterNames = ["balanceManager"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "generate_proof_as_owner",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function generateProofAsTrader(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::balance_manager::BalanceManager`,
    `${packageAddress}::balance_manager::TradeCap`
  ];
  const parameterNames = ["balanceManager", "tradeCap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "generate_proof_as_trader",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function deposit(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::balance_manager::BalanceManager`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${options.typeArguments[0]}>`
  ];
  const parameterNames = ["balanceManager", "coin"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "deposit",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function depositWithCap(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::balance_manager::BalanceManager`,
    `${packageAddress}::balance_manager::DepositCap`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${options.typeArguments[0]}>`
  ];
  const parameterNames = ["balanceManager", "depositCap", "coin"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "deposit_with_cap",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function withdrawWithCap(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::balance_manager::BalanceManager`,
    `${packageAddress}::balance_manager::WithdrawCap`,
    "u64"
  ];
  const parameterNames = ["balanceManager", "withdrawCap", "withdrawAmount"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "withdraw_with_cap",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function withdraw(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::balance_manager::BalanceManager`,
    "u64"
  ];
  const parameterNames = ["balanceManager", "withdrawAmount"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "withdraw",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function withdrawAll(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::balance_manager::BalanceManager`];
  const parameterNames = ["balanceManager"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "withdraw_all",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function validateProof(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::balance_manager::BalanceManager`,
    `${packageAddress}::balance_manager::TradeProof`
  ];
  const parameterNames = ["balanceManager", "proof"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "validate_proof",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function owner(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::balance_manager::BalanceManager`];
  const parameterNames = ["balanceManager"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "owner",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function id(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::balance_manager::BalanceManager`];
  const parameterNames = ["balanceManager"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "balance_manager",
    function: "id",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
export {
  BalanceEvent,
  BalanceKey,
  BalanceManager,
  BalanceManagerEvent,
  DepositCap,
  TradeCap,
  TradeProof,
  WithdrawCap,
  _new,
  balance,
  deposit,
  depositWithCap,
  generateProofAsOwner,
  generateProofAsTrader,
  id,
  mintDepositCap,
  mintTradeCap,
  mintWithdrawCap,
  newWithCustomOwner,
  newWithCustomOwnerAndCaps,
  newWithOwner,
  owner,
  revokeTradeCap,
  validateProof,
  withdraw,
  withdrawAll,
  withdrawWithCap
};
