// @mysten/deepbook-v3@0.26.6/dist/esm/contracts/deepbook/big_vector.js
import { MoveStruct } from "../utils/index.js";
import { bcs } from "@mysten/sui/bcs";
import * as object from "./deps/sui/object.js";
const $moduleName = "@deepbook/core::big_vector";
const BigVector = new MoveStruct({
  name: `${$moduleName}::BigVector`,
  fields: {
    id: object.UID,
    /** How deep the tree structure is. */
    depth: bcs.u8(),
    /**
     * Total number of elements that this vector contains, not including gaps in the
     * vector.
     */
    length: bcs.u64(),
    /** Max size of leaf nodes (counted in number of elements, `E`). */
    max_slice_size: bcs.u64(),
    /** Max size of interior nodes (counted in number of children). */
    max_fan_out: bcs.u64(),
    /** ID of the tree's root structure. Value of `NO_SLICE` means there's no root. */
    root_id: bcs.u64(),
    /** The last node ID that was allocated. */
    last_id: bcs.u64()
  }
});
function Slice(...typeParameters) {
  return new MoveStruct({
    name: `${$moduleName}::Slice<${typeParameters[0].name}>`,
    fields: {
      /** Previous node in the intrusive doubly-linked list data structure. */
      prev: bcs.u64(),
      /** Next node in the intrusive doubly-linked list data structure. */
      next: bcs.u64(),
      keys: bcs.vector(bcs.u128()),
      vals: bcs.vector(typeParameters[0])
    }
  });
}
const SliceRef = new MoveStruct({
  name: `${$moduleName}::SliceRef`,
  fields: {
    ix: bcs.u64()
  }
});
export {
  BigVector,
  Slice,
  SliceRef
};
