// @mysten/deepbook-v3@0.26.6/dist/esm/contracts/deepbook/constants.js
function currentVersion(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "current_version"
  });
}
function poolCreationFee(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "pool_creation_fee"
  });
}
function floatScaling(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "float_scaling"
  });
}
function floatScalingU128(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "float_scaling_u128"
  });
}
function maxU64(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "max_u64"
  });
}
function maxU128(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "max_u128"
  });
}
function noRestriction(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "no_restriction"
  });
}
function immediateOrCancel(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "immediate_or_cancel"
  });
}
function fillOrKill(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "fill_or_kill"
  });
}
function postOnly(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "post_only"
  });
}
function maxRestriction(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "max_restriction"
  });
}
function live(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "live"
  });
}
function partiallyFilled(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "partially_filled"
  });
}
function filled(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "filled"
  });
}
function canceled(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "canceled"
  });
}
function expired(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "expired"
  });
}
function selfMatchingAllowed(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "self_matching_allowed"
  });
}
function cancelTaker(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "cancel_taker"
  });
}
function cancelMaker(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "cancel_maker"
  });
}
function minPrice(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "min_price"
  });
}
function maxPrice(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "max_price"
  });
}
function phaseOutEpochs(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "phase_out_epochs"
  });
}
function defaultStakeRequired(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "default_stake_required"
  });
}
function half(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "half"
  });
}
function feeIsDeep(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "fee_is_deep"
  });
}
function deepUnit(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "deep_unit"
  });
}
function maxFills(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "max_fills"
  });
}
function maxOpenOrders(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "max_open_orders"
  });
}
function maxSliceSize(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "max_slice_size"
  });
}
function maxFanOut(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "max_fan_out"
  });
}
function feePenaltyMultiplier(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "fee_penalty_multiplier"
  });
}
function defaultEwmaAlpha(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "default_ewma_alpha"
  });
}
function defaultZScoreThreshold(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "default_z_score_threshold"
  });
}
function defaultAdditionalTakerFee(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "default_additional_taker_fee"
  });
}
function ewmaDfKey(options = {}) {
  const packageAddress = options.package ?? "@deepbook/core";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "constants",
    function: "ewma_df_key"
  });
}
export {
  cancelMaker,
  cancelTaker,
  canceled,
  currentVersion,
  deepUnit,
  defaultAdditionalTakerFee,
  defaultEwmaAlpha,
  defaultStakeRequired,
  defaultZScoreThreshold,
  ewmaDfKey,
  expired,
  feeIsDeep,
  feePenaltyMultiplier,
  fillOrKill,
  filled,
  floatScaling,
  floatScalingU128,
  half,
  immediateOrCancel,
  live,
  maxFanOut,
  maxFills,
  maxOpenOrders,
  maxPrice,
  maxRestriction,
  maxSliceSize,
  maxU128,
  maxU64,
  minPrice,
  noRestriction,
  partiallyFilled,
  phaseOutEpochs,
  poolCreationFee,
  postOnly,
  selfMatchingAllowed
};
