// @mysten/deepbook-v3@0.26.6/dist/esm/contracts/deepbook/governance.js
import { MoveStruct } from "../utils/index.js";
import { bcs } from "@mysten/sui/bcs";
import * as vec_map from "./deps/sui/vec_map.js";
import * as trade_params from "./trade_params.js";
const $moduleName = "@deepbook/core::governance";
const Proposal = new MoveStruct({
  name: `${$moduleName}::Proposal`,
  fields: {
    taker_fee: bcs.u64(),
    maker_fee: bcs.u64(),
    stake_required: bcs.u64(),
    votes: bcs.u64()
  }
});
const Governance = new MoveStruct({
  name: `${$moduleName}::Governance`,
  fields: {
    /** Tracks refreshes. */
    epoch: bcs.u64(),
    /** If Pool is whitelisted. */
    whitelisted: bcs.bool(),
    /** If Pool is stable or volatile. */
    stable: bcs.bool(),
    /** List of proposals for the current epoch. */
    proposals: vec_map.VecMap(bcs.Address, Proposal),
    /** Trade parameters for the current epoch. */
    trade_params: trade_params.TradeParams,
    /** Trade parameters for the next epoch. */
    next_trade_params: trade_params.TradeParams,
    /** All voting power from the current stakes. */
    voting_power: bcs.u64(),
    /** Quorum for the current epoch. */
    quorum: bcs.u64()
  }
});
const TradeParamsUpdateEvent = new MoveStruct({
  name: `${$moduleName}::TradeParamsUpdateEvent`,
  fields: {
    taker_fee: bcs.u64(),
    maker_fee: bcs.u64(),
    stake_required: bcs.u64()
  }
});
export {
  Governance,
  Proposal,
  TradeParamsUpdateEvent
};
