// @mysten/deepbook-v3@0.26.6/dist/esm/contracts/deepbook/math.js
import { normalizeMoveArguments } from "../utils/index.js";
function mul(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["u64", "u64"];
  const parameterNames = ["x", "y"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "math",
    function: "mul",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function mulU128(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["u128", "u128"];
  const parameterNames = ["x", "y"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "math",
    function: "mul_u128",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function mulRoundUp(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["u64", "u64"];
  const parameterNames = ["x", "y"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "math",
    function: "mul_round_up",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function div(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["u64", "u64"];
  const parameterNames = ["x", "y"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "math",
    function: "div",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function divU128(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["u128", "u128"];
  const parameterNames = ["x", "y"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "math",
    function: "div_u128",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function divRoundUp(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["u64", "u64"];
  const parameterNames = ["x", "y"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "math",
    function: "div_round_up",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function median(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["vector<u128>"];
  const parameterNames = ["v"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "math",
    function: "median",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function sqrt(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["u64", "u64"];
  const parameterNames = ["x", "precision"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "math",
    function: "sqrt",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function isPowerOfTen(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = ["u64"];
  const parameterNames = ["n"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "math",
    function: "is_power_of_ten",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
export {
  div,
  divRoundUp,
  divU128,
  isPowerOfTen,
  median,
  mul,
  mulRoundUp,
  mulU128,
  sqrt
};
