// @mysten/deepbook-v3@0.26.6/dist/esm/contracts/deepbook/order_query.js
import { MoveStruct, normalizeMoveArguments } from "../utils/index.js";
import { bcs } from "@mysten/sui/bcs";
import * as order from "./order.js";
const $moduleName = "@deepbook/core::order_query";
const OrderPage = new MoveStruct({
  name: `${$moduleName}::OrderPage`,
  fields: {
    orders: bcs.vector(order.Order),
    has_next_page: bcs.bool()
  }
});
function iterOrders(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::pool::Pool<${options.typeArguments[0]}, ${options.typeArguments[1]}>`,
    "0x0000000000000000000000000000000000000000000000000000000000000001::option::Option<u128>",
    "0x0000000000000000000000000000000000000000000000000000000000000001::option::Option<u128>",
    "0x0000000000000000000000000000000000000000000000000000000000000001::option::Option<u64>",
    "u64",
    "bool"
  ];
  const parameterNames = [
    "self",
    "startOrderId",
    "endOrderId",
    "minExpireTimestamp",
    "limit",
    "bids"
  ];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_query",
    function: "iter_orders",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function orders(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_query::OrderPage`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_query",
    function: "orders",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function hasNextPage(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_query::OrderPage`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_query",
    function: "has_next_page",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
export {
  OrderPage,
  hasNextPage,
  iterOrders,
  orders
};
