// @mysten/deepbook-v3@0.26.6/dist/esm/contracts/deepbook/order_info.js
import { MoveStruct, normalizeMoveArguments } from "../utils/index.js";
import { bcs } from "@mysten/sui/bcs";
import * as deep_price from "./deep_price.js";
import * as fill from "./fill.js";
const $moduleName = "@deepbook/core::order_info";
const OrderInfo = new MoveStruct({
  name: `${$moduleName}::OrderInfo`,
  fields: {
    pool_id: bcs.Address,
    order_id: bcs.u128(),
    balance_manager_id: bcs.Address,
    client_order_id: bcs.u64(),
    trader: bcs.Address,
    order_type: bcs.u8(),
    self_matching_option: bcs.u8(),
    price: bcs.u64(),
    is_bid: bcs.bool(),
    original_quantity: bcs.u64(),
    order_deep_price: deep_price.OrderDeepPrice,
    expire_timestamp: bcs.u64(),
    executed_quantity: bcs.u64(),
    cumulative_quote_quantity: bcs.u64(),
    fills: bcs.vector(fill.Fill),
    fee_is_deep: bcs.bool(),
    paid_fees: bcs.u64(),
    maker_fees: bcs.u64(),
    epoch: bcs.u64(),
    status: bcs.u8(),
    market_order: bcs.bool(),
    fill_limit_reached: bcs.bool(),
    order_inserted: bcs.bool(),
    timestamp: bcs.u64()
  }
});
const OrderFilled = new MoveStruct({
  name: `${$moduleName}::OrderFilled`,
  fields: {
    pool_id: bcs.Address,
    maker_order_id: bcs.u128(),
    taker_order_id: bcs.u128(),
    maker_client_order_id: bcs.u64(),
    taker_client_order_id: bcs.u64(),
    price: bcs.u64(),
    taker_is_bid: bcs.bool(),
    taker_fee: bcs.u64(),
    taker_fee_is_deep: bcs.bool(),
    maker_fee: bcs.u64(),
    maker_fee_is_deep: bcs.bool(),
    base_quantity: bcs.u64(),
    quote_quantity: bcs.u64(),
    maker_balance_manager_id: bcs.Address,
    taker_balance_manager_id: bcs.Address,
    timestamp: bcs.u64()
  }
});
const OrderPlaced = new MoveStruct({
  name: `${$moduleName}::OrderPlaced`,
  fields: {
    balance_manager_id: bcs.Address,
    pool_id: bcs.Address,
    order_id: bcs.u128(),
    client_order_id: bcs.u64(),
    trader: bcs.Address,
    price: bcs.u64(),
    is_bid: bcs.bool(),
    placed_quantity: bcs.u64(),
    expire_timestamp: bcs.u64(),
    timestamp: bcs.u64()
  }
});
const OrderExpired = new MoveStruct({
  name: `${$moduleName}::OrderExpired`,
  fields: {
    balance_manager_id: bcs.Address,
    pool_id: bcs.Address,
    order_id: bcs.u128(),
    client_order_id: bcs.u64(),
    trader: bcs.Address,
    price: bcs.u64(),
    is_bid: bcs.bool(),
    original_quantity: bcs.u64(),
    base_asset_quantity_canceled: bcs.u64(),
    timestamp: bcs.u64()
  }
});
function poolId(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "pool_id",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function orderId(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "order_id",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function balanceManagerId(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "balance_manager_id",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function clientOrderId(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "client_order_id",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function trader(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "trader",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function orderType(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "order_type",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function selfMatchingOption(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "self_matching_option",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function price(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "price",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function isBid(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "is_bid",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function originalQuantity(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "original_quantity",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function orderDeepPrice(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "order_deep_price",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function expireTimestamp(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "expire_timestamp",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function executedQuantity(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "executed_quantity",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function cumulativeQuoteQuantity(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "cumulative_quote_quantity",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function fills(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "fills",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function feeIsDeep(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "fee_is_deep",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function paidFees(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "paid_fees",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function makerFees(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "maker_fees",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function epoch(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "epoch",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function status(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "status",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function fillLimitReached(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "fill_limit_reached",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function orderInserted(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order_info::OrderInfo`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order_info",
    function: "order_inserted",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
export {
  OrderExpired,
  OrderFilled,
  OrderInfo,
  OrderPlaced,
  balanceManagerId,
  clientOrderId,
  cumulativeQuoteQuantity,
  epoch,
  executedQuantity,
  expireTimestamp,
  feeIsDeep,
  fillLimitReached,
  fills,
  isBid,
  makerFees,
  orderDeepPrice,
  orderId,
  orderInserted,
  orderType,
  originalQuantity,
  paidFees,
  poolId,
  price,
  selfMatchingOption,
  status,
  trader
};
