// @mysten/deepbook-v3@0.26.6/dist/esm/contracts/deepbook/order.js
import { MoveStruct, normalizeMoveArguments } from "../utils/index.js";
import { bcs } from "@mysten/sui/bcs";
import * as deep_price from "./deep_price.js";
const $moduleName = "@deepbook/core::order";
const Order = new MoveStruct({
  name: `${$moduleName}::Order`,
  fields: {
    balance_manager_id: bcs.Address,
    order_id: bcs.u128(),
    client_order_id: bcs.u64(),
    quantity: bcs.u64(),
    filled_quantity: bcs.u64(),
    fee_is_deep: bcs.bool(),
    order_deep_price: deep_price.OrderDeepPrice,
    epoch: bcs.u64(),
    status: bcs.u8(),
    expire_timestamp: bcs.u64()
  }
});
const OrderCanceled = new MoveStruct({
  name: `${$moduleName}::OrderCanceled`,
  fields: {
    balance_manager_id: bcs.Address,
    pool_id: bcs.Address,
    order_id: bcs.u128(),
    client_order_id: bcs.u64(),
    trader: bcs.Address,
    price: bcs.u64(),
    is_bid: bcs.bool(),
    original_quantity: bcs.u64(),
    base_asset_quantity_canceled: bcs.u64(),
    timestamp: bcs.u64()
  }
});
const OrderModified = new MoveStruct({
  name: `${$moduleName}::OrderModified`,
  fields: {
    balance_manager_id: bcs.Address,
    pool_id: bcs.Address,
    order_id: bcs.u128(),
    client_order_id: bcs.u64(),
    trader: bcs.Address,
    price: bcs.u64(),
    is_bid: bcs.bool(),
    previous_quantity: bcs.u64(),
    filled_quantity: bcs.u64(),
    new_quantity: bcs.u64(),
    timestamp: bcs.u64()
  }
});
function balanceManagerId(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order::Order`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order",
    function: "balance_manager_id",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function orderId(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order::Order`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order",
    function: "order_id",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function clientOrderId(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order::Order`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order",
    function: "client_order_id",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function quantity(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order::Order`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order",
    function: "quantity",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function filledQuantity(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order::Order`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order",
    function: "filled_quantity",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function feeIsDeep(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order::Order`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order",
    function: "fee_is_deep",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function orderDeepPrice(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order::Order`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order",
    function: "order_deep_price",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function epoch(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order::Order`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order",
    function: "epoch",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function status(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order::Order`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order",
    function: "status",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function expireTimestamp(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order::Order`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order",
    function: "expire_timestamp",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function price(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [`${packageAddress}::order::Order`];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "order",
    function: "price",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
export {
  Order,
  OrderCanceled,
  OrderModified,
  balanceManagerId,
  clientOrderId,
  epoch,
  expireTimestamp,
  feeIsDeep,
  filledQuantity,
  orderDeepPrice,
  orderId,
  price,
  quantity,
  status
};
