// @mysten/deepbook-v3@0.26.6/dist/esm/contracts/deepbook/registry.js
import { MoveStruct, normalizeMoveArguments } from "../utils/index.js";
import { bcs } from "@mysten/sui/bcs";
import * as object from "./deps/sui/object.js";
import * as versioned from "./deps/sui/versioned.js";
import * as vec_set from "./deps/sui/vec_set.js";
import * as bag from "./deps/sui/bag.js";
import * as type_name from "./deps/std/type_name.js";
const $moduleName = "@deepbook/core::registry";
const REGISTRY = new MoveStruct({
  name: `${$moduleName}::REGISTRY`,
  fields: {
    dummy_field: bcs.bool()
  }
});
const DeepbookAdminCap = new MoveStruct({
  name: `${$moduleName}::DeepbookAdminCap`,
  fields: {
    id: object.UID
  }
});
const Registry = new MoveStruct({
  name: `${$moduleName}::Registry`,
  fields: {
    id: object.UID,
    inner: versioned.Versioned
  }
});
const RegistryInner = new MoveStruct({
  name: `${$moduleName}::RegistryInner`,
  fields: {
    allowed_versions: vec_set.VecSet(bcs.u64()),
    pools: bag.Bag,
    treasury_address: bcs.Address
  }
});
const PoolKey = new MoveStruct({
  name: `${$moduleName}::PoolKey`,
  fields: {
    base: type_name.TypeName,
    quote: type_name.TypeName
  }
});
const StableCoinKey = new MoveStruct({
  name: `${$moduleName}::StableCoinKey`,
  fields: {
    dummy_field: bcs.bool()
  }
});
function setTreasuryAddress(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::registry::Registry`,
    "address",
    `${packageAddress}::registry::DeepbookAdminCap`
  ];
  const parameterNames = ["self", "treasuryAddress", "Cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "registry",
    function: "set_treasury_address",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function enableVersion(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::registry::Registry`,
    "u64",
    `${packageAddress}::registry::DeepbookAdminCap`
  ];
  const parameterNames = ["self", "version", "Cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "registry",
    function: "enable_version",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function disableVersion(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::registry::Registry`,
    "u64",
    `${packageAddress}::registry::DeepbookAdminCap`
  ];
  const parameterNames = ["self", "version", "Cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "registry",
    function: "disable_version",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function addStablecoin(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::registry::Registry`,
    `${packageAddress}::registry::DeepbookAdminCap`
  ];
  const parameterNames = ["self", "Cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "registry",
    function: "add_stablecoin",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function removeStablecoin(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::registry::Registry`,
    `${packageAddress}::registry::DeepbookAdminCap`
  ];
  const parameterNames = ["self", "Cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "registry",
    function: "remove_stablecoin",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function isStablecoin(options) {
  const packageAddress = options.package ?? "@deepbook/core";
  const argumentsTypes = [
    `${packageAddress}::registry::Registry`,
    "0x0000000000000000000000000000000000000000000000000000000000000001::type_name::TypeName"
  ];
  const parameterNames = ["self", "stableType"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "registry",
    function: "is_stablecoin",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
export {
  DeepbookAdminCap,
  PoolKey,
  REGISTRY,
  Registry,
  RegistryInner,
  StableCoinKey,
  addStablecoin,
  disableVersion,
  enableVersion,
  isStablecoin,
  removeStablecoin,
  setTreasuryAddress
};
