// @mysten/deepbook-v3@0.26.6/dist/esm/contracts/deepbook/state.js
import { MoveStruct } from "../utils/index.js";
import { bcs } from "@mysten/sui/bcs";
import * as table from "./deps/sui/table.js";
import * as history from "./history.js";
import * as governance from "./governance.js";
import * as balances from "./balances.js";
const $moduleName = "@deepbook/core::state";
const State = new MoveStruct({
  name: `${$moduleName}::State`,
  fields: {
    accounts: table.Table,
    history: history.History,
    governance: governance.Governance
  }
});
const StakeEvent = new MoveStruct({
  name: `${$moduleName}::StakeEvent`,
  fields: {
    pool_id: bcs.Address,
    balance_manager_id: bcs.Address,
    epoch: bcs.u64(),
    amount: bcs.u64(),
    stake: bcs.bool()
  }
});
const ProposalEvent = new MoveStruct({
  name: `${$moduleName}::ProposalEvent`,
  fields: {
    pool_id: bcs.Address,
    balance_manager_id: bcs.Address,
    epoch: bcs.u64(),
    taker_fee: bcs.u64(),
    maker_fee: bcs.u64(),
    stake_required: bcs.u64()
  }
});
const VoteEvent = new MoveStruct({
  name: `${$moduleName}::VoteEvent`,
  fields: {
    pool_id: bcs.Address,
    balance_manager_id: bcs.Address,
    epoch: bcs.u64(),
    from_proposal_id: bcs.option(bcs.Address),
    to_proposal_id: bcs.Address,
    stake: bcs.u64()
  }
});
const RebateEventV2 = new MoveStruct({
  name: `${$moduleName}::RebateEventV2`,
  fields: {
    pool_id: bcs.Address,
    balance_manager_id: bcs.Address,
    epoch: bcs.u64(),
    claim_amount: balances.Balances
  }
});
const RebateEvent = new MoveStruct({
  name: `${$moduleName}::RebateEvent`,
  fields: {
    pool_id: bcs.Address,
    balance_manager_id: bcs.Address,
    epoch: bcs.u64(),
    claim_amount: bcs.u64()
  }
});
export {
  ProposalEvent,
  RebateEvent,
  RebateEventV2,
  StakeEvent,
  State,
  VoteEvent
};
