// @mysten/deepbook-v3@0.26.6/dist/cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  Account: () => import_bcs.Account,
  BalanceManagerContract: () => import_balanceManager.BalanceManagerContract,
  Balances: () => import_bcs.Balances,
  ConfigurationError: () => import_errors.ConfigurationError,
  DEEP_SCALAR: () => import_config2.DEEP_SCALAR,
  DeepBookAdminContract: () => import_deepbookAdmin.DeepBookAdminContract,
  DeepBookClient: () => import_client.DeepBookClient,
  DeepBookConfig: () => import_config.DeepBookConfig,
  DeepBookContract: () => import_deepbook.DeepBookContract,
  DeepBookError: () => import_errors.DeepBookError,
  ErrorMessages: () => import_errors.ErrorMessages,
  FLOAT_SCALAR: () => import_config2.FLOAT_SCALAR,
  FlashLoanContract: () => import_flashLoans.FlashLoanContract,
  GAS_BUDGET: () => import_config2.GAS_BUDGET,
  GovernanceContract: () => import_governance.GovernanceContract,
  MAX_TIMESTAMP: () => import_config2.MAX_TIMESTAMP,
  MarginAdminContract: () => import_marginAdmin.MarginAdminContract,
  MarginMaintainerContract: () => import_marginMaintainer.MarginMaintainerContract,
  MarginManagerContract: () => import_marginManager.MarginManagerContract,
  MarginPoolContract: () => import_marginPool.MarginPoolContract,
  MarginTPSLContract: () => import_marginTPSL.MarginTPSLContract,
  Order: () => import_bcs.Order,
  OrderDeepPrice: () => import_bcs.OrderDeepPrice,
  OrderType: () => import_types.OrderType,
  POOL_CREATION_FEE_DEEP: () => import_config2.POOL_CREATION_FEE_DEEP,
  PRICE_INFO_OBJECT_MAX_AGE_MS: () => import_config2.PRICE_INFO_OBJECT_MAX_AGE_MS,
  PoolProxyContract: () => import_poolProxy.PoolProxyContract,
  ResourceNotFoundError: () => import_errors.ResourceNotFoundError,
  SelfMatchingOptions: () => import_types.SelfMatchingOptions,
  SuiPriceServiceConnection: () => import_pyth.SuiPriceServiceConnection,
  SuiPythClient: () => import_pyth.SuiPythClient,
  ValidationError: () => import_errors.ValidationError,
  VecSet: () => import_bcs.VecSet,
  validateAddress: () => import_validation.validateAddress,
  validateNonEmptyArray: () => import_validation.validateNonEmptyArray,
  validateNonNegativeNumber: () => import_validation.validateNonNegativeNumber,
  validatePositiveNumber: () => import_validation.validatePositiveNumber,
  validateRange: () => import_validation.validateRange,
  validateRequired: () => import_validation.validateRequired
});
module.exports = __toCommonJS(index_exports);
var import_client = require("./client.js");
var import_config = require("./utils/config.js");
var import_balanceManager = require("./transactions/balanceManager.js");
var import_deepbook = require("./transactions/deepbook.js");
var import_deepbookAdmin = require("./transactions/deepbookAdmin.js");
var import_flashLoans = require("./transactions/flashLoans.js");
var import_governance = require("./transactions/governance.js");
var import_marginAdmin = require("./transactions/marginAdmin.js");
var import_marginMaintainer = require("./transactions/marginMaintainer.js");
var import_marginManager = require("./transactions/marginManager.js");
var import_marginPool = require("./transactions/marginPool.js");
var import_poolProxy = require("./transactions/poolProxy.js");
var import_marginTPSL = require("./transactions/marginTPSL.js");
var import_pyth = require("./pyth/pyth.js");
var import_bcs = require("./types/bcs.js");
var import_types = require("./types/index.js");
var import_config2 = require("./utils/config.js");
var import_errors = require("./utils/errors.js");
var import_validation = require("./utils/validation.js");
