// @mysten/deepbook-v3@0.26.6/dist/cjs/transactions/marginManager.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __accessCheck = (obj, member, msg) => member.has(obj) || __typeError("Cannot " + msg);
var __privateGet = (obj, member, getter) => (__accessCheck(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd = (obj, member, value) => member.has(obj) ? __typeError("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet = (obj, member, value, setter) => (__accessCheck(obj, member, "write to private field"), setter ? setter.call(obj, value) : member.set(obj, value), value);
var marginManager_exports = {};
__export(marginManager_exports, {
  MarginManagerContract: () => MarginManagerContract
});
module.exports = __toCommonJS(marginManager_exports);
var import_transactions = require("@mysten/sui/transactions");
var _config;
class MarginManagerContract {
  /**
   * @param {DeepBookConfig} config Configuration for MarginManagerContract
   */
  constructor(config) {
    __privateAdd(this, _config);
    /**
     * @description Create a new margin manager
     * @param {string} poolKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.newMarginManager = (poolKey) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::new`,
        arguments: [
          tx.object(pool.address),
          tx.object(__privateGet(this, _config).REGISTRY_ID),
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Create a new margin manager with an initializer
     * @param {string} poolKey The key to identify the pool
     * @returns A function that takes a Transaction object
     */
    this.newMarginManagerWithInitializer = (poolKey) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const [manager, initializer] = tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::new_with_initializer`,
        arguments: [
          tx.object(pool.address),
          tx.object(__privateGet(this, _config).REGISTRY_ID),
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
      return { manager, initializer };
    };
    /**
     * @description Share a margin manager
     * @param {string} poolKey The key to identify the pool
     * @param {TransactionArgument} manager The margin manager to share
     * @param {TransactionArgument} initializer The initializer for the manager
     * @returns A function that takes a Transaction object
     */
    this.shareMarginManager = (poolKey, manager, initializer) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::share`,
        arguments: [manager, initializer],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Deposit base into a margin manager
     * @param {string} managerKey The key to identify the manager
     * @param {number} amount The amount to deposit
     * @returns A function that takes a Transaction object
     */
    this.depositBase = (managerKey, amount) => (tx) => {
      const manager = __privateGet(this, _config).getMarginManager(managerKey);
      const pool = __privateGet(this, _config).getPool(manager.poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const coin = (0, import_transactions.coinWithBalance)({
        type: baseCoin.type,
        balance: amount * baseCoin.scalar
      });
      tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::deposit`,
        arguments: [
          tx.object(manager.address),
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(baseCoin.priceInfoObjectId),
          tx.object(quoteCoin.priceInfoObjectId),
          coin,
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type, baseCoin.type]
      });
    };
    /**
     * @description Deposit quote into a margin manager
     * @param {string} managerKey The key to identify the manager
     * @param {number} amount The amount to deposit
     * @returns A function that takes a Transaction object
     */
    this.depositQuote = (managerKey, amount) => (tx) => {
      const manager = __privateGet(this, _config).getMarginManager(managerKey);
      const pool = __privateGet(this, _config).getPool(manager.poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const coin = (0, import_transactions.coinWithBalance)({
        type: quoteCoin.type,
        balance: amount * quoteCoin.scalar
      });
      tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::deposit`,
        arguments: [
          tx.object(manager.address),
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(baseCoin.priceInfoObjectId),
          tx.object(quoteCoin.priceInfoObjectId),
          coin,
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Deposit deep into a margin manager
     * @param {string} managerKey The key to identify the manager
     * @param {number} amount The amount to deposit
     * @returns A function that takes a Transaction object
     */
    this.depositDeep = (managerKey, amount) => (tx) => {
      const manager = __privateGet(this, _config).getMarginManager(managerKey);
      const pool = __privateGet(this, _config).getPool(manager.poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const deepCoin = __privateGet(this, _config).getCoin("DEEP");
      const coin = (0, import_transactions.coinWithBalance)({
        type: deepCoin.type,
        balance: amount * deepCoin.scalar
      });
      tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::deposit`,
        arguments: [
          tx.object(manager.address),
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(baseCoin.priceInfoObjectId),
          tx.object(quoteCoin.priceInfoObjectId),
          coin,
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type, deepCoin.type]
      });
    };
    /**
     * @description Withdraw base from a margin manager
     * @param {string} managerKey The key to identify the manager
     * @param {number} amount The amount to withdraw
     * @returns A function that takes a Transaction object
     */
    this.withdrawBase = (managerKey, amount) => (tx) => {
      const manager = __privateGet(this, _config).getMarginManager(managerKey);
      const pool = __privateGet(this, _config).getPool(manager.poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const baseMarginPool = __privateGet(this, _config).getMarginPool(pool.baseCoin);
      const quoteMarginPool = __privateGet(this, _config).getMarginPool(pool.quoteCoin);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::withdraw`,
        arguments: [
          tx.object(manager.address),
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(baseMarginPool.address),
          tx.object(quoteMarginPool.address),
          tx.object(baseCoin.priceInfoObjectId),
          tx.object(quoteCoin.priceInfoObjectId),
          tx.object(pool.address),
          tx.pure.u64(amount * baseCoin.scalar),
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type, baseCoin.type]
      });
    };
    /**
     * @description Withdraw quote from a margin manager
     * @param {string} managerKey The key to identify the manager
     * @param {number} amount The amount to withdraw
     * @returns A function that takes a Transaction object
     */
    this.withdrawQuote = (managerKey, amount) => (tx) => {
      const manager = __privateGet(this, _config).getMarginManager(managerKey);
      const pool = __privateGet(this, _config).getPool(manager.poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const baseMarginPool = __privateGet(this, _config).getMarginPool(pool.baseCoin);
      const quoteMarginPool = __privateGet(this, _config).getMarginPool(pool.quoteCoin);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::withdraw`,
        arguments: [
          tx.object(manager.address),
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(baseMarginPool.address),
          tx.object(quoteMarginPool.address),
          tx.object(baseCoin.priceInfoObjectId),
          tx.object(quoteCoin.priceInfoObjectId),
          tx.object(pool.address),
          tx.pure.u64(amount * quoteCoin.scalar),
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Withdraw deep from a margin manager
     * @param {string} managerKey The key to identify the manager
     * @param {number} amount The amount to withdraw
     * @returns A function that takes a Transaction object
     */
    this.withdrawDeep = (managerKey, amount) => (tx) => {
      const manager = __privateGet(this, _config).getMarginManager(managerKey);
      const pool = __privateGet(this, _config).getPool(manager.poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const deepCoin = __privateGet(this, _config).getCoin("DEEP");
      const baseMarginPool = __privateGet(this, _config).getMarginPool(pool.baseCoin);
      const quoteMarginPool = __privateGet(this, _config).getMarginPool(pool.quoteCoin);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::withdraw`,
        arguments: [
          tx.object(manager.address),
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(baseMarginPool.address),
          tx.object(quoteMarginPool.address),
          tx.object(baseCoin.priceInfoObjectId),
          tx.object(quoteCoin.priceInfoObjectId),
          tx.object(pool.address),
          tx.pure.u64(amount * deepCoin.scalar),
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type, deepCoin.type]
      });
    };
    /**
     * @description Borrow base from a margin manager
     * @param {string} managerKey The key to identify the manager
     * @param {number} amount The amount to borrow
     * @returns A function that takes a Transaction object
     */
    this.borrowBase = (managerKey, amount) => (tx) => {
      const manager = __privateGet(this, _config).getMarginManager(managerKey);
      const pool = __privateGet(this, _config).getPool(manager.poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const baseMarginPool = __privateGet(this, _config).getMarginPool(pool.baseCoin);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::borrow_base`,
        arguments: [
          tx.object(manager.address),
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(baseMarginPool.address),
          tx.object(baseCoin.priceInfoObjectId),
          tx.object(quoteCoin.priceInfoObjectId),
          tx.object(pool.address),
          tx.pure.u64(amount * baseCoin.scalar),
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Borrow quote from a margin manager
     * @param {string} managerKey The key to identify the manager
     * @param {number} amount The amount to borrow
     * @returns A function that takes a Transaction object
     */
    this.borrowQuote = (managerKey, amount) => (tx) => {
      const manager = __privateGet(this, _config).getMarginManager(managerKey);
      const pool = __privateGet(this, _config).getPool(manager.poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const quoteMarginPool = __privateGet(this, _config).getMarginPool(pool.quoteCoin);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::borrow_quote`,
        arguments: [
          tx.object(manager.address),
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(quoteMarginPool.address),
          tx.object(baseCoin.priceInfoObjectId),
          tx.object(quoteCoin.priceInfoObjectId),
          tx.object(pool.address),
          tx.pure.u64(amount * quoteCoin.scalar),
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Repay base from a margin manager
     * @param {string} managerKey The key to identify the manager
     * @param {number} amount The amount to repay
     * @returns A function that takes a Transaction object
     */
    this.repayBase = (managerKey, amount) => (tx) => {
      const manager = __privateGet(this, _config).getMarginManager(managerKey);
      const pool = __privateGet(this, _config).getPool(manager.poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const baseMarginPool = __privateGet(this, _config).getMarginPool(pool.baseCoin);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::repay_base`,
        arguments: [
          tx.object(manager.address),
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(baseMarginPool.address),
          tx.object.option({
            type: "u64",
            value: amount ? tx.pure.u64(amount * baseCoin.scalar) : null
          }),
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Repay quote from a margin manager
     * @param {string} managerKey The key to identify the manager
     * @param {number} amount The amount to repay
     * @returns A function that takes a Transaction object
     */
    this.repayQuote = (managerKey, amount) => (tx) => {
      const manager = __privateGet(this, _config).getMarginManager(managerKey);
      const pool = __privateGet(this, _config).getPool(manager.poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const quoteMarginPool = __privateGet(this, _config).getMarginPool(pool.quoteCoin);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::repay_quote`,
        arguments: [
          tx.object(manager.address),
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(quoteMarginPool.address),
          tx.object.option({
            type: "u64",
            value: amount ? tx.pure.u64(amount * quoteCoin.scalar) : null
          }),
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Liquidate a margin manager
     * @param {string} managerAddress The address of the manager to liquidate
     * @param {string} poolKey The key to identify the pool
     * @param {boolean} debtIsBase Whether the debt is in base
     * @param {TransactionArgument} repayCoin The coin to repay
     * @returns A function that takes a Transaction object
     */
    this.liquidate = (managerAddress, poolKey, debtIsBase, repayCoin) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const baseMarginPool = __privateGet(this, _config).getMarginPool(pool.baseCoin);
      const quoteMarginPool = __privateGet(this, _config).getMarginPool(pool.quoteCoin);
      const marginPool = debtIsBase ? baseMarginPool : quoteMarginPool;
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::liquidate`,
        arguments: [
          tx.object(managerAddress),
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(baseCoin.priceInfoObjectId),
          tx.object(quoteCoin.priceInfoObjectId),
          tx.object(marginPool.address),
          tx.object(pool.address),
          repayCoin,
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Set the referral for a margin manager (DeepBookPoolReferral)
     * @param {string} managerKey The key to identify the margin manager
     * @param {string} referral The referral (DeepBookPoolReferral) to set
     * @returns A function that takes a Transaction object
     */
    this.setMarginManagerReferral = (managerKey, referral) => (tx) => {
      const manager = __privateGet(this, _config).getMarginManager(managerKey);
      const pool = __privateGet(this, _config).getPool(manager.poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::set_margin_manager_referral`,
        arguments: [tx.object(manager.address), tx.object(referral)],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Unset the referral for a margin manager
     * @param {string} managerKey The key to identify the margin manager
     * @param {string} poolKey The key of the pool to unset the referral for
     * @returns A function that takes a Transaction object
     */
    this.unsetMarginManagerReferral = (managerKey, poolKey) => (tx) => {
      const manager = __privateGet(this, _config).getMarginManager(managerKey);
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::unset_margin_manager_referral`,
        arguments: [tx.object(manager.address), tx.pure.id(pool.address)],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    // === Read-Only Functions ===
    /**
     * @description Get the owner address of a margin manager
     * @param {string} poolKey The key to identify the pool
     * @param {string} marginManagerId The ID of the margin manager
     * @returns A function that takes a Transaction object
     */
    this.ownerByPoolKey = (poolKey, marginManagerId) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::owner`,
        arguments: [tx.object(marginManagerId)],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Get the DeepBook pool ID associated with a margin manager
     * @param {string} poolKey The key to identify the pool
     * @param {string} marginManagerId The ID of the margin manager
     * @returns A function that takes a Transaction object
     */
    this.deepbookPool = (poolKey, marginManagerId) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::deepbook_pool`,
        arguments: [tx.object(marginManagerId)],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Get the margin pool ID (if any) associated with a margin manager
     * @param {string} poolKey The key to identify the pool
     * @param {string} marginManagerId The ID of the margin manager
     * @returns A function that takes a Transaction object
     */
    this.marginPoolId = (poolKey, marginManagerId) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::margin_pool_id`,
        arguments: [tx.object(marginManagerId)],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Get borrowed shares for both base and quote assets
     * @param {string} poolKey The key to identify the pool
     * @param {string} marginManagerId The ID of the margin manager
     * @returns A function that takes a Transaction object
     */
    this.borrowedShares = (poolKey, marginManagerId) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::borrowed_shares`,
        arguments: [tx.object(marginManagerId)],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Get borrowed base shares
     * @param {string} poolKey The key to identify the pool
     * @param {string} marginManagerId The ID of the margin manager
     * @returns A function that takes a Transaction object
     */
    this.borrowedBaseShares = (poolKey, marginManagerId) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::borrowed_base_shares`,
        arguments: [tx.object(marginManagerId)],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Get borrowed quote shares
     * @param {string} poolKey The key to identify the pool
     * @param {string} marginManagerId The ID of the margin manager
     * @returns A function that takes a Transaction object
     */
    this.borrowedQuoteShares = (poolKey, marginManagerId) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::borrowed_quote_shares`,
        arguments: [tx.object(marginManagerId)],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Check if margin manager has base asset debt
     * @param {string} poolKey The key to identify the pool
     * @param {string} marginManagerId The ID of the margin manager
     * @returns A function that takes a Transaction object
     */
    this.hasBaseDebt = (poolKey, marginManagerId) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::has_base_debt`,
        arguments: [tx.object(marginManagerId)],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Get the balance manager ID for a margin manager
     * @param {string} poolKey The key to identify the pool
     * @param {string} marginManagerId The ID of the margin manager
     * @returns A function that takes a Transaction object
     */
    this.balanceManager = (poolKey, marginManagerId) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::balance_manager`,
        arguments: [tx.object(marginManagerId)],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Calculate assets (base and quote) for a margin manager
     * @param {string} poolKey The key to identify the pool
     * @param {string} marginManagerId The ID of the margin manager
     * @returns A function that takes a Transaction object
     */
    this.calculateAssets = (poolKey, marginManagerId) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::calculate_assets`,
        arguments: [tx.object(marginManagerId), tx.object(pool.address)],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Calculate debts (base and quote) for a margin manager
     * @param {string} poolKey The key to identify the pool
     * @param {string} coinKey The key to identify the debt coin (base or quote)
     * @param {string} marginManagerId The ID of the margin manager
     * @returns A function that takes a Transaction object
     */
    this.calculateDebts = (poolKey, coinKey, marginManagerId) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const debtCoin = __privateGet(this, _config).getCoin(coinKey);
      const marginPool = __privateGet(this, _config).getMarginPool(coinKey);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::calculate_debts`,
        arguments: [tx.object(marginManagerId), tx.object(marginPool.address), tx.object.clock()],
        typeArguments: [baseCoin.type, quoteCoin.type, debtCoin.type]
      });
    };
    /**
     * @description Get comprehensive state information for a margin manager
     * @param {string} poolKey The key to identify the pool
     * @param {string} marginManagerId The ID of the margin manager
     * @returns A function that takes a Transaction object
     * @returns Returns (manager_id, deepbook_pool_id, risk_ratio, base_asset, quote_asset,
     *                   base_debt, quote_debt, base_pyth_price, base_pyth_decimals,
     *                   quote_pyth_price, quote_pyth_decimals)
     */
    this.managerState = (poolKey, marginManagerId) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      const baseMarginPool = __privateGet(this, _config).getMarginPool(pool.baseCoin);
      const quoteMarginPool = __privateGet(this, _config).getMarginPool(pool.quoteCoin);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::manager_state`,
        arguments: [
          tx.object(marginManagerId),
          tx.object(__privateGet(this, _config).MARGIN_REGISTRY_ID),
          tx.object(baseCoin.priceInfoObjectId),
          tx.object(quoteCoin.priceInfoObjectId),
          tx.object(pool.address),
          tx.object(baseMarginPool.address),
          tx.object(quoteMarginPool.address),
          tx.object.clock()
        ],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Get the base asset balance of a margin manager
     * @param {string} poolKey The key to identify the pool
     * @param {string} marginManagerId The ID of the margin manager
     * @returns A function that takes a Transaction object
     */
    this.baseBalance = (poolKey, marginManagerId) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::base_balance`,
        arguments: [tx.object(marginManagerId)],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Get the quote asset balance of a margin manager
     * @param {string} poolKey The key to identify the pool
     * @param {string} marginManagerId The ID of the margin manager
     * @returns A function that takes a Transaction object
     */
    this.quoteBalance = (poolKey, marginManagerId) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::quote_balance`,
        arguments: [tx.object(marginManagerId)],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    /**
     * @description Get the DEEP token balance of a margin manager
     * @param {string} poolKey The key to identify the pool
     * @param {string} marginManagerId The ID of the margin manager
     * @returns A function that takes a Transaction object
     */
    this.deepBalance = (poolKey, marginManagerId) => (tx) => {
      const pool = __privateGet(this, _config).getPool(poolKey);
      const baseCoin = __privateGet(this, _config).getCoin(pool.baseCoin);
      const quoteCoin = __privateGet(this, _config).getCoin(pool.quoteCoin);
      return tx.moveCall({
        target: `${__privateGet(this, _config).MARGIN_PACKAGE_ID}::margin_manager::deep_balance`,
        arguments: [tx.object(marginManagerId)],
        typeArguments: [baseCoin.type, quoteCoin.type]
      });
    };
    __privateSet(this, _config, config);
  }
}
_config = new WeakMap();
